<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyVIPER (VIPER Analysis in Python for single-cell RNASeq) &mdash; pyviper 1.0.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pyviper package" href="pyviper.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating metacells for reverse-engineering of ARACNe gene regulatory networks</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><img width="50" alt="image" src="https://github.com/alevax/pyviper/assets/92543296/3b1f14ab-68a1-49ab-b5b2-aa4f214ad995"> pyVIPER (VIPER Analysis in Python for single-cell RNASeq)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="img-width-50-alt-image-src-https-github-com-alevax-pyviper-assets-92543296-3b1f14ab-68a1-49ab-b5b2-aa4f214ad995-pyviper-viper-analysis-in-python-for-single-cell-rnaseq">
<h1><img width="50" alt="image" src="https://github.com/alevax/pyviper/assets/92543296/3b1f14ab-68a1-49ab-b5b2-aa4f214ad995"> pyVIPER (VIPER Analysis in Python for single-cell RNASeq)<a class="headerlink" href="#img-width-50-alt-image-src-https-github-com-alevax-pyviper-assets-92543296-3b1f14ab-68a1-49ab-b5b2-aa4f214ad995-pyviper-viper-analysis-in-python-for-single-cell-rnaseq" title="Permalink to this heading"></a></h1>
<!-- [![PyPI](https://img.shields.io/pypi/v/viper-in-python?logo=PyPI)](https://pypi.org/project/viper-in-python) -->
<p><a class="reference external" href="https://pypi.org/project/viper-in-python"><img alt="PyPI" src="https://img.shields.io/badge/pypi-stable-darkgreeen?logo=PyPI" /></a>
<a class="reference external" href="https://opensource.org/licenses/MIT"><img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg" /></a>
<a class="reference external" href="https://pepy.tech/project/viper-in-python"><img alt="Downloads" src="https://static.pepy.tech/badge/viper-in-python" /></a></p>
<p>This package enables network-based protein activity estimation on Python.
It provides also interfaces for scanpy (single-cell RNASeq analysis in Python).
Functions are partly transplanted from R package <a class="reference external" href="https://www.bioconductor.org/packages/release/bioc/html/viper.html">viper</a> and the R package <a class="reference external" href="https://github.com/califano-lab/NaRnEA">NaRnEA</a>.</p>
<p>The user-friendly documentation is available at: https://alevax.github.io/pyviper/index.html.</p>
<div align="center">
    <img style="width: 80%;" alt="viper_visualized" src="https://github.com/user-attachments/assets/44d230cd-7eb3-4cc1-8e27-4cb9d49c269d">
</div>
<hr class="docutils" />
<div class="section" id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">scanpy</span></code> for single cell pipeline</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pandas</span></code> and <code class="docutils literal notranslate"><span class="pre">anndata</span></code> for data computing and storage.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">numpy</span></code> and <code class="docutils literal notranslate"><span class="pre">scipy</span></code>  for scientific computation.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">joblib</span></code> for parallel computing</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tqdm</span></code> show progress bar</p></li>
</ul>
<p>If you are using a version of <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> &lt;1.9.3, it is also advisable to downgrade <code class="docutils literal notranslate"><span class="pre">pandas</span></code> to (&gt;=1.3.0 &amp; &lt;2.0), due to <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> incompatibility (<a class="reference external" href="https://github.com/scverse/scanpy/issues/2564">issue</a>)</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h2>
<div class="section" id="pypi">
<h3>pypi<a class="headerlink" href="#pypi" title="Permalink to this heading"></a></h3>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>viper-in-python
</pre></div>
</div>
</div>
<div class="section" id="local">
<h3>local<a class="headerlink" href="#local" title="Permalink to this heading"></a></h3>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/alevax/pyviper/
<span class="nb">cd</span><span class="w"> </span>pyviper
pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
</pre></div>
</div>
</div>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">pyviper</span>

<span class="c1"># Load sample data</span>
<span class="n">ges</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">read_text</span><span class="p">(</span><span class="s2">&quot;test/unit_tests/test_1/test_1_inputs/LNCaPWT_gExpr_GES.tsv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>

<span class="c1"># Load network</span>
<span class="n">network</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">load</span><span class="o">.</span><span class="n">msigdb_regulon</span><span class="p">(</span><span class="s2">&quot;h&quot;</span><span class="p">)</span>

<span class="c1"># Translate sample data from ensembl to gene names</span>
<span class="n">pyviper</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">ges</span><span class="p">,</span> <span class="n">desired_format</span> <span class="o">=</span> <span class="s2">&quot;human_symbol&quot;</span><span class="p">)</span>

<span class="c1">## Filter targets in the interactome</span>
<span class="n">network</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">ges</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>

<span class="c1"># Compute regulon activities</span>
<span class="c1">## area</span>
<span class="n">activity</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">gex_data</span><span class="o">=</span><span class="n">ges</span><span class="p">,</span> <span class="n">interactome</span><span class="o">=</span><span class="n">network</span><span class="p">,</span> <span class="n">enrichment</span><span class="o">=</span><span class="s2">&quot;area&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">activity</span><span class="o">.</span><span class="n">to_df</span><span class="p">())</span>

<span class="c1">## narnea</span>
<span class="n">activity</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">gex_data</span><span class="o">=</span><span class="n">ges</span><span class="p">,</span> <span class="n">interactome</span><span class="o">=</span><span class="n">network</span><span class="p">,</span> <span class="n">enrichment</span><span class="o">=</span><span class="s2">&quot;narnea&quot;</span><span class="p">,</span> <span class="n">eset_filter</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">activity</span><span class="o">.</span><span class="n">to_df</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="section" id="tutorials">
<h2>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://github.com/alevax/pyviper/blob/main/Tutorials/Tutorial-1.ipynb">Analyzing scRNA-seq data at the Protein Activity Level</a></p></li>
<li><p><a class="reference external" href="https://github.com/alevax/pyviper/blob/main/Tutorials/Tutorial-2.ipynb">Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></p></li>
<li><p><a class="reference external" href="https://github.com/alevax/pyviper/blob/main/Tutorials/Tutorial-3.ipynb">Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></p></li>
</ol>
</div>
<div class="section" id="structure-and-rationale">
<h2>Structure and rationale<a class="headerlink" href="#structure-and-rationale" title="Permalink to this heading"></a></h2>
<p>The main functions available from <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.viper</span></code>: “pyviper” function for Virtual Inference of Protein Activity by Enriched Regulon Analysis (VIPER). The function allows using 2 enrichment algorithms, aREA and (matrix)-NaRnEA (see below).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.aREA</span></code>: computes <a class="reference external" href="https://www.nature.com/articles/ng.3593">aREA</a> (analytic rank-based enrichment analysis) and meta-aREA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.NaRnEA</span></code>: computes <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2021.05.20.445002v5">matrix-NaRnEA</a>, a vectorized, implementation of <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">NaRnEA</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pp.translate</span></code>: for translating between species (i.e. mouse vs human) and between ensembl, entrez and gene symbols.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.tl.path_enr</span></code>: computes pathway enrichment</p></li>
</ul>
<p>Other notable functions include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.tl.OncoMatch</span></code>: computes <a class="reference external" href="https://doi.org/10.1158/2159-8290.CD-22-0342">OncoMatch</a>, an algorithm to assess the activity conservation of MR proteins between two sets of samples (e.g. validate GEMMs as effective models of human samples)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pp.stouffer</span></code>: computes signatures on a cluster-by-cluster basis using Cluster integration method for pathway enrichment</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pp.viper_similarity</span></code>: computes the <a class="reference external" href="https://s3.jcloud.sjtu.edu.cn/899a892efef34b1b944a19981040f55b-oss01/bioconductor/3.14/bioc/vignettes/viper/inst/doc/viper.pdf">similarity</a> between VIPER signatures</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pp.repr_metacells</span></code>: compute representative metacells (e.g. for ARACNe) using our method to maximize unique sample usage and minimize resampling (users can specify depth, percent data usage, etc).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pp.repr_subsample</span></code>: select a representative subsample of data using our method to ensure a widely distributed sampling.</p></li>
</ul>
<p>Additionally, the following submodules are available:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.load</span></code>: submodule containing several utility functions useful for different analyses, including <code class="docutils literal notranslate"><span class="pre">load_msigdb_regulon</span></code>, <code class="docutils literal notranslate"><span class="pre">load_TFs</span></code> etc</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.pl</span></code>: submodule containing pyviper-wrappers for <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> plotting</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.tl</span></code>: submodule containing pyviper-wrappers for <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> data transformation</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pyviper.config</span></code>: submodule allowing users to specify current species and filepaths for regulators</p></li>
</ul>
<p>Last, a new <code class="docutils literal notranslate"><span class="pre">Interactome</span></code> class allows users to load and interrogate ARACNe- and SCENIC-inferred gene regulatory networks.</p>
</div>
<div class="section" id="contact">
<h2>Contact<a class="headerlink" href="#contact" title="Permalink to this heading"></a></h2>
<p>Please, report any issues that you experience through this repository <a class="reference external" href="https://github.com/alevax/pyviper/issues">“Issues”</a>.</p>
<p>For any other info or queries please write to Alessandro Vasciaveo (av2729&#64;cumc.columbia.edu)</p>
</div>
<div class="section" id="license">
<h2>License<a class="headerlink" href="#license" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pyviper</span></code> is distributed under a MIT License (see <a class="reference external" href="https://github.com/alevax/pyviper/blob/main/LICENSE">LICENSE</a>).</p>
</div>
<div class="section" id="citation">
<h2>Citation<a class="headerlink" href="#citation" title="Permalink to this heading"></a></h2>
<p>If you used pyVIPER in your publication, please cite our work here:</p>
<p>Wang, A.L.E., Lin, Z., Zanella, L., Vlahos, L., Girotto, M.A., Zafar, A., … &amp; Vasciaveo, A. (2024). pyVIPER: A fast and scalable Python package for rank-based enrichment analysis of single-cell RNASeq data. <em>bioRxiv</em>, 2024-08. doi: https://doi.org/10.1101/2024.08.25.609585.</p>
<p><em>Manuscript in review</em></p>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#pyviper-area">pyviper.aREA</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#pyviper-narnea">pyviper.NaRnEA</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#module-pyviper.config">pyviper.config</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#pyviper-interactome">pyviper.Interactome</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#module-pyviper.load">pyviper.load</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#module-pyviper.pl">pyviper.pl</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#module-pyviper.pp">pyviper.pp</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#module-pyviper.tl">pyviper.tl</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyviper.html#pyviper-viper">pyviper.viper</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Install-PyVIPER">Install PyVIPER</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Import-modules">Import modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Step-1.-Load-a-gene-expression-signature-for-single-cells">Step 1. Load a gene expression signature for single-cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Step-2.-Load-an-inspect-a-lineage-specific-gene-regulatory-network">Step 2. Load an inspect a lineage-specific gene regulatory network</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Step-3.-Convert-the-gene-expression-signature-into-a-protein-activity-matrix-using-VIPER">Step 3. Convert the gene expression signature into a protein activity matrix using VIPER</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Step-4.-Analyze-single-cells-at-the-Protein-Activity-level">Step 4. Analyze single-cells at the Protein Activity level</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-1.html#Key-takeaways">Key takeaways</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Install-PyVIPER">Install PyVIPER</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Import-modules">Import modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">Step 1. Load a gene expression matrix and associated metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Step-2.-Preprocess-and-generate-a-gene-expression-signature-at-the-single-cell-level">Step 2. Preprocess and generate a gene expression signature at the single-cell level</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Step-3.-Load-multiple-ARACNe-inferred-gene-regulatory-networks">Step 3. Load multiple ARACNe-inferred gene regulatory networks</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Step-4.-Analyze-single-cells-at-the-Protein-Activity-level">Step 4. Analyze single-cells at the Protein Activity level</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Pathway-enrichment-analysis">Pathway enrichment analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-2.html#Key-takeaways">Key takeaways</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating metacells for reverse-engineering of ARACNe gene regulatory networks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#Install-Pyviper">Install Pyviper</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#Import-modules">Import modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">Step 1. Load a gene expression matrix and associated metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#Step-2.-Perform-basic-preprocessing-and-clustering">Step 2. Perform basic preprocessing and clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#3.-Generate-Metacells">3. Generate Metacells</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial-3.html#Key-takeaways">Key takeaways</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pyviper.html" class="btn btn-neutral float-right" title="pyviper package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>