<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyviper.aREA package &mdash; pyviper 1.0.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">pyviper.aREA package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pyviper.aREA.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="pyviper-area-package">
<h1>pyviper.aREA package<a class="headerlink" href="#pyviper-area-package" title="Permalink to this heading"></a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="module-pyviper.aREA.aREA_classic">
<span id="pyviper-area-area-classic-module"></span><h2>pyviper.aREA.aREA_classic module<a class="headerlink" href="#module-pyviper.aREA.aREA_classic" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_classic.aREA_classic">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_classic.</span></span><span class="sig-name descname"><span class="pre">aREA_classic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_classic.aREA_classic" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the analytical ranked enrichment analysis (aREA)
function.</p>
<p>It is the basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – The interactome object.</p></li>
<li><p><strong>layer</strong> – The layer in the anndata object to use as the gene expression input
(default = None).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.aREA.aREA_meta">
<span id="pyviper-area-area-meta-module"></span><h2>pyviper.aREA.aREA_meta module<a class="headerlink" href="#module-pyviper.aREA.aREA_meta" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_meta.aREA">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_meta.</span></span><span class="sig-name descname"><span class="pre">aREA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mvws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_meta.aREA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the Analytical Ranked Enrichment Analysis (aREA)[1]
function.</p>
<p>It is the original basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommended to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A consistent number of
targets allows regulators to have NES scores that are comparable to one
another. A regulator that has more targets than others will have “boosted”
NES scores, such that they cannot be compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – (A) Number indicating either the exponent score for the metaViper weights.
These are only applicable when enrichment = ‘area’ and are not used when
enrichment = ‘narnea’. Roughly, a lower number (e.g. 1) results in
networks being treated as a consensus network (useful for multiple
networks of the same celltype with the same epigenetics), while a higher
number (e.g. 10) results in networks being treated as separate (useful
for multiple networks of different celltypes with different epigenetics).
(B) The name of a column in gex_data that contains the manual assignments
of samples to networks using list position or network names.
(C) “auto”: assign samples to networks based on how well each
network allows for sample enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>[1] Alvarez, M. J., Shen, Y., Giorgi, F. M., Lachmann, A., Ding, B. B., Ye,</em></p></li>
<li><p><em>B. H., &amp; Califano, A. (2016). Functional characterization of somatic</em></p></li>
<li><p><em>mutations in cancer using network-based inference of protein activity.</em></p></li>
<li><p><em>Nature genetics, 48(8), 838-847.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.aREA">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pyviper.aREA" title="Permalink to this heading"></a></h2>
</div>
</div>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>