<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyviper.NaRnEA package &mdash; pyviper 1.0.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating metacells for reverse-engineering of ARACNe gene regulatory networks</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">pyviper.NaRnEA package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pyviper.NaRnEA.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="pyviper-narnea-package">
<h1>pyviper.NaRnEA package<a class="headerlink" href="#pyviper-narnea-package" title="Permalink to this heading"></a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="pyviper-narnea-narnea-classic-module">
<h2>pyviper.NaRnEA.NaRnEA_classic module<a class="headerlink" href="#pyviper-narnea-narnea-classic-module" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="module-pyviper.NaRnEA.NaRnEA_meta">
<span id="pyviper-narnea-narnea-meta-module"></span><h2>pyviper.NaRnEA.NaRnEA_meta module<a class="headerlink" href="#module-pyviper.NaRnEA.NaRnEA_meta" title="Permalink to this heading"></a></h2>
<p>Allows the individual to infer normalized enrichment scores and proportional
enrichment scores from gene expression data using the Nonparametric
Analytical Rank-based Enrichment Analysis (NaRnEA)[1] function.</p>
<p>NaRnEA is an updated basis for the VIPER (Virtual Inference of
Protein-activity by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommend to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A regulator that has more
targets than others will have “boosted” NES scores, such that they cannot be
compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">param gex_data</dt>
<dd class="field-odd"><p>Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p>
</dd>
<dt class="field-even">param interactome</dt>
<dd class="field-even"><p>An object of class Interactome.</p>
</dd>
<dt class="field-odd">param layer</dt>
<dd class="field-odd"><p>The layer in the anndata object to use as the gene expression input.</p>
</dd>
<dt class="field-even">type layer</dt>
<dd class="field-even"><p>default: None</p>
</dd>
<dt class="field-odd">param eset_filter</dt>
<dd class="field-odd"><p>Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p>
</dd>
<dt class="field-even">type eset_filter</dt>
<dd class="field-even"><p>default: False</p>
</dd>
<dt class="field-odd">param min_targets</dt>
<dd class="field-odd"><p>The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p>
</dd>
<dt class="field-even">type min_targets</dt>
<dd class="field-even"><p>default: 30</p>
</dd>
<dt class="field-odd">param verbose</dt>
<dd class="field-odd"><p>Whether extended output about the progress of the algorithm should be
given.</p>
</dd>
<dt class="field-even">type verbose</dt>
<dd class="field-even"><p>default: True</p>
</dd>
<dt class="field-odd">returns</dt>
<dd class="field-odd"><p><strong>A dictionary containing :class:`~numpy.ndarray` containing NES values (key</strong></p>
</dd>
<dt class="field-even">rtype</dt>
<dd class="field-even"><p>‘nes’) and PES values (key: ‘pes’).</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>[1] Griffin, A. T., Vlahos, L. J., Chiuzan, C., &amp; Califano, A. (2023). NaRnEA: An Information Theoretic Framework for Gene Set Analysis. Entropy, 25(3), 542.</p>
</div>
<div class="section" id="module-pyviper.NaRnEA">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pyviper.NaRnEA" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.NaRnEA.NaRnEA">
<span class="sig-prename descclassname"><span class="pre">pyviper.NaRnEA.</span></span><span class="sig-name descname"><span class="pre">NaRnEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.NaRnEA.NaRnEA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores and proportional
enrichment scores from gene expression data using the Nonparametric
Analytical Rank-based Enrichment Analysis (NaRnEA)[1] function.</p>
<p>NaRnEA is an updated basis for the VIPER (Virtual Inference of
Protein-activity by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommend to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A regulator that has more
targets than others will have “boosted” NES scores, such that they cannot be
compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>layer</strong> (<em>default: None</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>eset_filter</strong> (<em>default: False</em>) – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>min_targets</strong> (<em>default: 30</em>) – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>verbose</strong> (<em>default: True</em>) – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>A dictionary containing :class:`~numpy.ndarray` containing NES values (key</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>‘nes’) and PES values (key: ‘pes’).</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>[1] Griffin, A. T., Vlahos, L. J., Chiuzan, C., &amp; Califano, A. (2023). NaRnEA: An Information Theoretic Framework for Gene Set Analysis. Entropy, 25(3), 542.</p>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>