<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyviper package &mdash; pyviper 1.0.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" href="Tutorial-1.html" />
    <link rel="prev" title="pyVIPER (VIPER Analysis in Python for single-cell RNASeq)" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">pyviper package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.config">pyviper.config</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.interactome">pyviper.interactome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.pl">pyviper.pl</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.pp">pyviper.pp</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.tl">pyviper.tl</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.aREA.aREA_classic">pyviper.aREA.aREA_classic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.aREA.aREA_meta">pyviper.aREA.aREA_meta</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.NaRnEA.NaRnEA_classic">pyviper.NaRnEA.NaRnEA_classic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.NaRnEA.NaRnEA_meta">pyviper.NaRnEA.NaRnEA_meta</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">pyviper package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pyviper.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="pyviper-package">
<h1>pyviper package<a class="headerlink" href="#pyviper-package" title="Permalink to this heading"></a></h1>
<div class="section" id="module-pyviper.config">
<span id="pyviper-config"></span><h2>pyviper.config<a class="headerlink" href="#module-pyviper.config" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.config.set_regulators_filepath">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_filepath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.config.set_regulators_filepath" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to use a custom list of regulatory proteins instead of the
default ones within pyVIPER’s data folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group</strong> – A group of regulatory proteins of either: “tfs”, “cotfs”, “sig” or “surf”.</p></li>
<li><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p></li>
<li><p><strong>new_filepath</strong> – The new filepath that should be used to retrieve these sets of proteins.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.config.set_regulators_species_to_use">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_species_to_use</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.config.set_regulators_species_to_use" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to specify which species they are currently studying, so the
correct sets of regulatory proteins will be used during analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id0">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_filepath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id0" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to use a custom list of regulatory proteins instead of the
default ones within pyVIPER’s data folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group</strong> – A group of regulatory proteins of either: “tfs”, “cotfs”, “sig” or “surf”.</p></li>
<li><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p></li>
<li><p><strong>new_filepath</strong> – The new filepath that should be used to retrieve these sets of proteins.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id1">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_species_to_use</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id1" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to specify which species they are currently studying, so the
correct sets of regulatory proteins will be used during analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.interactome">
<span id="pyviper-interactome"></span><h2>pyviper.interactome<a class="headerlink" href="#module-pyviper.interactome" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyviper.interactome.</span></span><span class="sig-name descname"><span class="pre">Interactome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.copy" title="Permalink to this definition"></a></dt>
<dd><p>Create a copy of this Interactome object.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An object of <a class="reference internal" href="#pyviper.interactome.Interactome" title="pyviper.interactome.Interactome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interactome</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.filter_regulators">
<span class="sig-name descname"><span class="pre">filter_regulators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regulators_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regulators_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.filter_regulators" title="Permalink to this definition"></a></dt>
<dd><p>Filter regulators by choosing by name or by group which ones you intend
to keep and which ones you intend to remove from this Interactome.</p>
<p>Note that the names of regulator that belong to the groups “tfs”, “cotfs”,
“sig” and “surf” will be sourced via the paths specified in pyviper.config.
To update these paths, use the pyviper.config.set_regulators_filepath
function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – This should be either:
(1) An array or list containing the names of specific regulators you
wish to keep in the network. When left as None, this parameter is
not used to filter.
(2) An array or list containing a group or groups of regulators that
you wish to keep in the network. These groups should be one of the
following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – This should be either:
(1) An array or list containing the names of specific regulators you
wish to remove from the network. When left as None, this parameter
is not used to filter.
(2) An array or list containing a group or groups of regulators that
you wish to remove from the network. These groups should be one of the
following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of regulators removed during filtering</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.filter_targets">
<span class="sig-name descname"><span class="pre">filter_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">targets_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targets_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.filter_targets" title="Permalink to this definition"></a></dt>
<dd><p>Filter targets by choosing by name which ones you intend to keep and
which ones you intend remove from this Interactome.</p>
<p>When working with an anndata object or a gene expression array, it is
highly recommended to filter the unPruned network before pruning. This
is to ensure the pruned network contains a consistent number of targets
per regulator regulator, all of which exist within gex_data. A regulator
that has more targets than others will have “boosted” NES scores, such
that they cannot be compared to those with fewer targets.
For example, with an anndata object named gex_data, one may is suggested
to do:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – An array containing the names of targets you wish to keep in the
network. When left as None, this parameter is not used to filter.</p></li>
<li><p><strong>(</strong><strong>default</strong> – An array containing the names of targets you wish to remove from the
network. When left as None, this parameter is not used to filter.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of targets removed during filtering</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_reg">
<span class="sig-name descname"><span class="pre">get_reg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regName</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_reg" title="Permalink to this definition"></a></dt>
<dd><p>Get the rows of the net_table where the regulator is regName.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regName</strong> – The name of a regulator in this Interactome.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_reg_names">
<span class="sig-name descname"><span class="pre">get_reg_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_reg_names" title="Permalink to this definition"></a></dt>
<dd><p>Get an array of all unique regulators in this Interactome.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An array of strings of <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_target_names">
<span class="sig-name descname"><span class="pre">get_target_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_target_names" title="Permalink to this definition"></a></dt>
<dd><p>Get a set of the unique targets in this Interactome</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A 1D NumPy array.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.ic_mat">
<span class="sig-name descname"><span class="pre">ic_mat</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.ic_mat" title="Permalink to this definition"></a></dt>
<dd><p>Get the DataFrame of all the likelihood values. Targets are in the rows,
while Regulators are in the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.icp_vec">
<span class="sig-name descname"><span class="pre">icp_vec</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.icp_vec" title="Permalink to this definition"></a></dt>
<dd><p>Get the vector containing the proportion of the “Interaction Confidence”
(IC) score for each interaction in a network, relative to the maximum IC
score in the network. This vector is generated by taking each individual
regulon in the newtork and calculating the likelihood index proportion
to all interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An array of <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.integrate">
<span class="sig-name descname"><span class="pre">integrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_likelihoods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.integrate" title="Permalink to this definition"></a></dt>
<dd><p>Integrate this Interactome object with one or more other Interacome
objects to create a consensus network. In general, this should be done
when interactome objects have the same epigenetics (e.g. due to being
made from different datasets of same celltype). MetaVIPER should be used
instead when you have multiple interactomes with different epigenetics
(e.g. due to being made of data with different celltypes).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>network_list</strong> – A single object or a list of objects of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>normalize_likelihoods</em>) – An array containing weights for each network being integrated. The
first weight corresponds to this network, while the others correspond
to those in the network list in order. If None, equal weights are
used.</p></li>
<li><p><strong>(</strong><strong>default</strong> – An extra operation that can performed after the integration
operation where within each regulator, likelihood values are ranked
and scaled from 0 to 1.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.mor_mat">
<span class="sig-name descname"><span class="pre">mor_mat</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.mor_mat" title="Permalink to this definition"></a></dt>
<dd><p>Get the DataFrame of all the correlation values. Targets are in the rows,
while Regulators are in the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.prune">
<span class="sig-name descname"><span class="pre">prune</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eliminate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.prune" title="Permalink to this definition"></a></dt>
<dd><p>Prune the Interactome by eliminating extra targets from regulators and,
with eliminate = True, remove regulators with too few targets from the
network. Note that by ensuring the pruned networks contains the same
number of targets for each regulator, NES scores are comparable. If one
regulator has more targest than another, than its NES score will be
“boosted” and they cannot be compared against each other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The maximum number of targets that each regulon is allowed.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulon is required.</p></li>
<li><p><strong>(</strong><strong>default</strong> – If eliminate = True, then any regulators with fewer targets than
max_targets will be removed from the network. In other words, after
pruning, all regulators will have exactly max_targets number of
targets. This essentially sets min_targets equal to max_targets
and ensures all NES scores are comparable with aREA.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of targets and regulators removed during pruning</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the Interactome object to one’s disk. If saved as “csv” or
“tsv”, just the interactome.net_table will be saved. If saved as
“pkl”, the whole interactome object will be saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> – A filepath to one’s disk to store the Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>output_type</em>) – If None, the output_type will be inferred from the file_path.
Otherwise, specify “csv”, “tsv” or “pkl”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.size" title="Permalink to this definition"></a></dt>
<dd><p>Get the the number of regulators in this Interactome.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.translate_regulators">
<span class="sig-name descname"><span class="pre">translate_regulators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">desired_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.translate_regulators" title="Permalink to this definition"></a></dt>
<dd><p>Translate the regulators of the Interactome. The current name format of
the regulators should be one of the following:</p>
<blockquote>
<div><p>mouse_symbol, mouse_ensembl, mouse_entrez, human_symbol, human_ensembl or human_entrez</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>desired_format</strong> – Desired format can be one of four strings: “mouse_symbol”,
“mouse_ensembl”, “mouse_entrez”, “human_symbol”, “human_ensembl” or “human_entrez”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – Report the number of regulators successfully and unsucessfully translated</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.translate_targets">
<span class="sig-name descname"><span class="pre">translate_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">desired_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.translate_targets" title="Permalink to this definition"></a></dt>
<dd><p>Translate the targets of the Interactome.  The current name format of
the targets should be one of the following:</p>
<blockquote>
<div><p>mouse_symbol, mouse_ensembl, mouse_entrez, human_symbol, human_ensembl or human_entrez</p>
</div></blockquote>
<p>It is recommended to do this before pruning to ensure a consistent number
of targets because if targets do not have a translation, they will be
deleted, resulting in different numbers of targets in a pruned
interactome that once had consistent number of targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>desired_format</strong> – Desired format can be one of four strings: “mouse_symbol”,
“mouse_ensembl”, “mouse_entrez”, “human_symbol”, “human_ensembl” or “human_entrez”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – Report the number of targets successfully and unsucessfully translated</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pyviper.pl">
<span id="pyviper-pl"></span><h2>pyviper.pl<a class="headerlink" href="#module-pyviper.pl" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.__get_stored_uns_data_and_prep_to_plot">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">__get_stored_uns_data_and_prep_to_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uns_data_slot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uns_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.__get_stored_uns_data_and_prep_to_plot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.pca">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.pca" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.pca.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[‘X_pca’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[‘X_pca’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.pca function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.umap">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.umap" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.umap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[‘X_umap’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[‘X_umap’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.umap function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.tsne">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">tsne</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.tsne" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.tsne.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[‘X_tsne’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[‘X_tsne’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.tsne function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.diffmap">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">diffmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.diffmap" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.diffmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[‘X_diffmap’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[‘X_diffmap’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.diffmap function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.draw_graph">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">draw_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.draw_graph" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.draw_graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[‘X_draw_graph_fa’] or adata.obsm[‘X_draw_graph_fr’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[‘X_draw_graph_fa’] or adata.obsm[‘X_draw_graph_fr’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.draw_graph function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.spatial">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">spatial</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.spatial" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.spatial.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.uns[‘spatial’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.uns[‘spatial’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.spatial function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.embedding">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">embedding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.embedding" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.embedding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>basis</strong> – The name of the represenation in adata.obsm that should be used for plotting.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’] on adata.obsm[basis].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[basis].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.embedding function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.embedding_density">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">embedding_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.embedding_density" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.embedding_density.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – The name of the represenation in adata.obsm that should be used for plotting.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘gex_data’] on adata.obsm[basis].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’] on adata.obsm[basis].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.embedding_density function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.heatmap">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.heatmap" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.heatmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.heatmap function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.dotplot">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">dotplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.dotplot" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.dotplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.dotplot function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.tracksplot">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">tracksplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.tracksplot" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.tracksplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.tracksplot function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.violin">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">violin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.violin" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.violin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.violin function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.stacked_violin">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">stacked_violin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.stacked_violin" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.stacked_violin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.stacked_violin function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.matrixplot">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">matrixplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.matrixplot" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.matrixplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.matrixplot function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.clustermap">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">clustermap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.clustermap" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.clustermap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.clustermap function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.ranking">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">ranking</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.ranking" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.ranking.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.ranking function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pl.dendrogram">
<span class="sig-prename descclassname"><span class="pre">pyviper.pl.</span></span><span class="sig-name descname"><span class="pre">dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_gex_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_stored_pax_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pl.dendrogram" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.pl.dendrogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>plot_stored_pax_data</em>) – Plot adata.uns[‘gex_data’].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Plot adata.uns[‘pax_data’].</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.pl.dendrogram function.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A plot of <code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code>.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.pp">
<span id="pyviper-pp"></span><h2>pyviper.pp<a class="headerlink" href="#module-pyviper.pp" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.rank_norm">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">rank_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NUM_FUN=&lt;function</span> <span class="pre">_median&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DEM_FUN=&lt;function</span> <span class="pre">_mad_from_R&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.rank_norm" title="Permalink to this definition"></a></dt>
<dd><p>Compute a double rank normalization on an anndata, np.array, or pd.DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Data stored in an anndata object, np.array or pd.DataFrame.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – The first function to be applied across each column.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The second function to be applied across each column.</p></li>
<li><p><strong>(</strong><strong>default</strong> – For an anndata input, the layer to use. When None, the input layer is
anndata.X.</p></li>
<li><p><strong>(</strong><strong>default</strong> – For an anndata input, the name of the layer where to store. When None,
this is anndata.X.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to return a rank-transformed copy (True) or to instead transform
the original input (False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>When copy = False, saves the input data as a double rank transformed version.</em></p></li>
<li><p><em>When copy = True, return a double rank transformed version of the input data.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.stouffer">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">stouffer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'stouffer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_pvals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_iters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.stouffer" title="Permalink to this definition"></a></dt>
<dd><p>Compute a stouffer signature on each of your clusters in an anndata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata
object, or a pandas dataframe containing input data.</p></li>
<li><p><strong>obs_column_name</strong> – The name of the column of observations in adata to use as clusters, or a
cluster vector corresponding to observations.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – The layer to use as input data to compute the signatures.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.uns to store the stouffer signatures.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to compute a p-value for each score to return in the results.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The number of iterations to use to compute a null model to assess the
p-values of each of the stouffer scores.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to provide additional output during the execution of the function.</p></li>
<li><p><strong>(</strong><strong>default</strong> – If True, returns the stouffer signature in a pd.DataFrame. If False,
stores it in adata.var[key_added].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>When return_as_df is False, adds the cluster stouffer signatures to</em></p></li>
<li><p><em>adata.var[key_added]. When return_as_df is True, returns as pd.DataFrame.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.spearman">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">spearman</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'stouffer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_pvals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">null_iters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.spearman" title="Permalink to this definition"></a></dt>
<dd><p>Compute spearman correlation between each gene product and the cluster
centroids along with the statistical significance for each of your clusters
in an anndata object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata
object, or a pandas dataframe containing input data.</p></li>
<li><p><strong>pca_slot</strong> – The slot in adata.obsm where a PCA is stored.</p></li>
<li><p><strong>obs_column_name</strong> – The name of the column of observations in adata to use as clusters, or a
cluster vector corresponding to observations.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – The layer to use as input data to compute the correlation.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.uns to store the spearman correlation.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to compute a p-value for each score to return in the results.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The number of iterations to use to compute a null model to assess the
p-values of each of the spearman scores.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to provide additional output during the execution of the function.</p></li>
<li><p><strong>(</strong><strong>default</strong> – If True, returns the spearman signature in a pd.DataFrame. If False,
stores it in adata.var[key_added].</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>When return_as_df is False, adds the cluster spearman correlation to</em></p></li>
<li><p><em>adata.var[key_added]. When return_as_df is True, returns as pd.DataFrame.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.viper_similarity">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">viper_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[4,</span> <span class="pre">2]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['two-sided',</span> <span class="pre">'greater',</span> <span class="pre">'less']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viper_similarity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.viper_similarity" title="Permalink to this definition"></a></dt>
<dd><p>Compute the similarity between the columns of a VIPER-predicted activity or
gene expression matrix. While following the same concept as the two-tail
Gene Set Enrichment Analysis (GSEA)[1], it is based on the aREA algorithm[2].</p>
<p>If ws is a single number, weighting is performed using an exponential function.
If ws is a 2 numbers vector, weighting is performed with a symmetric sigmoid
function using the first element as inflection point and the second as trend.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – An anndata.AnnData containing protein activity (NES), where rows are
observations/samples (e.g. cells or groups) and columns are features
(e.g. proteins or pathways).</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – Optional number of top regulators to consider for computing the similarity</p></li>
<li><p><strong>(</strong><strong>default</strong> – Number indicating the weighting exponent for the signature, or vector of
2 numbers indicating the inflection point and the value corresponding to
a weighting score of .1 for a sigmoid transformation, only used if nn is
ommited.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Character string indicating whether the most active (greater), less
active (less) or both tails (two.sided) of the signature should be used
for computing the similarity.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The layer to use as input data to compute the signatures.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from the input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The name of the slot in the adata.obsp to store the output.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Saves a signature-based distance numpy.ndarray in adata.obsp[key_added].</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>[1] Julio, M. K. -d. et al. Regulation of extra-embryonic endoderm stem cell
differentiation by Nodal and Cripto signaling. Development 138, 3885-3895 (2011).
[2] Alvarez, M. J., Shen, Y., Giorgi, F. M., Lachmann, A., Ding, B. B., Ye, B. H.,
&amp; Califano, A. (2016). Functional characterization of somatic mutations in
cancer using network-based inference of protein activity. Nature genetics,
48(8), 838-847.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.aracne3_to_regulon">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">aracne3_to_regulon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">net_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anno</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MI_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regul_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_MI_per_regulon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.aracne3_to_regulon" title="Permalink to this definition"></a></dt>
<dd><p>Process an output from ARACNe3 to return a pd.DataFrame describing a gene
regulatory network with suitable columns for conversion to an object of the
Interactome class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>net_file</strong> – A string containing the path to the ARACNe3 output</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>normalize_MI_per_regulon</em>) – Whether to passt a pd.DataFrame instead of the path</p></li>
<li><p><strong>(</strong><strong>default</strong> – Gene ID annotation</p></li>
<li><p><strong>(</strong><strong>default</strong> – Threshold on Mutual Information (MI) to select the regulators and target pairs</p></li>
<li><p><strong>(</strong><strong>default</strong> – Number of (top) targets to include in each regulon</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to normalize MI values each regulon by the maximum value</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>A pd.DataFrame containing an ARACNe3-inferred gene regulatory network with the</em></p></li>
<li><p><strong>following 4 columns</strong> (<em>“regulator”, “target”, “mor” (mode of regulation) and “likelihood”.</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.nes_to_pval">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">nes_to_pval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_tail</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neg_log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.nes_to_pval" title="Permalink to this definition"></a></dt>
<dd><p>Transform VIPER-computed NES into p-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata
object, or a pandas dataframe containing input data, where rows are
observations/samples (e.g. cells or groups) and columns are features
(e.g. proteins or pathways).</p></li>
<li><p><strong>layer</strong> (<em>(</em><em>default: None</em><em>)</em>) – Entry of layers to tranform.</p></li>
<li><p><strong>key_added</strong> (<em>(</em><em>default: None</em><em>)</em>) – Name of layer to save result in a new layer instead of adata.X.</p></li>
<li><p><strong>lower_tail</strong> (<em>default</em><em> (</em><em>True</em><em>)</em>) – If <cite>True</cite> (default), probabilities are P(X &lt;= x)
If <cite>False</cite>, probabilities are P(X &gt; x)</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – If <cite>True</cite>, returns adjusted p values using FDR Benjamini-Hochberg procedure.
If <cite>False</cite>, does not adjust p values</p></li>
<li><p><strong>(</strong><strong>default</strong> – axis along which to perform the p-value correction (Used only if the input is a pd.DataFrame).
Possible values are 0 or 1.</p></li>
<li><p><strong>neg_log</strong> (<em>(</em><em>default: False</em><em>)</em>) – Whether to transform VIPER-computed NES into -log10(p-value).</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Saves the input data as a transformed version. If key_added is specified,</em></p></li>
<li><p><em>saves the results in adata.layers[key_added].</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.repr_subsample">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">repr_subsample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'repr_subsample'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eliminate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.repr_subsample" title="Permalink to this definition"></a></dt>
<dd><p>A tool for create a subsample of the input data such it is well
representative of all the populations within the input data rather than
being a random sample. This is accomplished by pairing samples together in
an iterative fashion until the desired sample size is reached.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – An anndata object containing a distance object in adata.obsp.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – The slot in adata.obsm where the PCA object is stored. One way of
generating this object is with sc.pp.pca.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The size of the representative subsample</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to trim down adata to the subsample (True) or leave the
subsample as an annotation in adata.obs[key_added].</p></li>
<li><p><strong>(</strong><strong>default</strong> – The random seed used when taking samples of the data.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to provide runtime information.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The number of cores to use for the analysis. Using more than 1 core
(multicore) speeds up the analysis.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>When copy is False, saves the subsample annotation in adata.var[key_added].</em></p></li>
<li><p><em>When copy is True, return an anndata with this annotation. When eliminate is</em></p></li>
<li><p><em>True, modify the adata by subsetting it down to the subsample.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.pp.repr_metacells">
<span class="sig-prename descclassname"><span class="pre">pyviper.pp.</span></span><span class="sig-name descname"><span class="pre">repr_metacells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'corr_dist'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_min_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cells_per_metacell</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_median_depth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc_data_to_use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc_incl_data_reused</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'metacells'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.pp.repr_metacells" title="Permalink to this definition"></a></dt>
<dd><p>A tool for create a representative selection of metacells from the data that
aims to maximize reusing samples from the data, while simultaneously
ensuring that all neighbors are close to the metacell they construct.
When using this function, exactly two of the following parameters must be
set: size, min_median_depth or n_cells_per_metacell, perc_data_to_use or
perc_incl_data_reused.
Note that min_median_depth and n_cells_per_metacell cannot both be set
at the same time, since they directly relate (e.g. higher n_cells_per_metacell
means more neighbors are used to construct a single metacell, meaning each
metacell will have more counts, resulting in a higher median depth).
Note that perc_data_to_use and perc_incl_data_reused cannot both be set
at the same time, since they directly relate (e.g. higher perc_data_to_use
means you include more data, which means it’s more likely to reuse more
data, resulting in a higher perc_incl_data_reused).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – An anndata object containing a distance object in adata.obsp.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – A pandas DataFrame or AnnData object of unnormalized gene expression
counts that has the same samples in the same order as that of adata.
If counts are left as None, adata must have counts stored in adata.raw.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.obsm where the PCA object is stored. One way of
generating this object is with sc.pp.pca.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.obsp where the distance object is stored. One way of
generating this object is with pyviper.pp.corr_distance.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.obs where cluster labels are stored. Cluster-specific
metacells will be generated using the same parameters with the results
for each cluster being stored separately in adata.uns.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in adata.obs where a score used to determine and filter cell
quality are stored (e.g. silhouette score).</p></li>
<li><p><strong>(</strong><strong>default</strong> – The score from adata.obs[score_slot] that a cell must have at minimum to
be used for metacell construction (e.g. 0.25 is the rule of thumb for
silhouette score).</p></li>
<li><p><strong>(</strong><strong>default</strong> – A specific number of metacells to generate. If set to None,
perc_data_to_use or perc_incl_data_reused can be used to specify the size
when n_cells_per_metacell or min_median_depth is given.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The number of cells that should be used to generate single metacell.
Note that this parameter and min_median_depth cannot both be set as
they directly relate: e.g. higher n_cells_per_metacell leads to higher
min_median_depth. If left as None, perc_data_to_use or
perc_incl_data_reused can be used to specify n_cells_per_metacell when
size is given.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The desired minimum median depth for the metacells (indirectly specifies
n_cells_per_metacell). The default is set to 10000 as this is recommend
by PISCES[1]. Note that this parameter and n_cells_per_metacell cannot
both be set as they directly relate: e.g. higher min_median_depth leads
to higher n_cells_per_metacell.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The percent of the total amount of provided samples that will be used in
the creation of metacells. Note that this parameter and
perc_incl_data_reused cannot both be set as they directly relate: e.g.
higher perc_data_to_use leads to higher perc_incl_data_reused.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The percent of samples that are included in the creation of metacells
that will be reused (i.e. used in more than one metacell). Note that this
parameter and perc_data_to_use cannot both be set as they directly relate:
e.g. higher perc_incl_data_reused leads to higher perc_data_to_use.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The random seed used when taking samples of the data.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The name of the slot in the adata.uns to store the output.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to provide runtime information and quality statistics.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The number of cores to use for the analysis. Using more than 1 core
(multicore) speeds up the analysis.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Saves the metacells as a pandas dataframe in adata.uns[key_added]. Attributes</em></p></li>
<li><p><em>that contain parameters for and statistics about the construction of the</em></p></li>
<li><p><em>metacells are stored in adata.uns[key_added].attrs. Set copy = True to</em></p></li>
<li><p><em>return a new AnnData object.</em></p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>Obradovic, A., Vlahos, L., Laise, P., Worley, J., Tan, X., Wang, A., &amp;</em></p></li>
<li><p><strong>Califano, A. (2021). PISCES</strong> (<em>A pipeline for the systematic, protein activity</em>)</p></li>
<li><p><em>-based analysis of single cell RNA sequencing data. bioRxiv, 6, 22.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.tl">
<span id="pyviper-tl"></span><h2>pyviper.tl<a class="headerlink" href="#module-pyviper.tl" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.tl.pca">
<span class="sig-prename descclassname"><span class="pre">pyviper.tl.</span></span><span class="sig-name descname"><span class="pre">pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.tl.pca" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.tl.pca.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>filter_by_feature_groups</em>) – The layer to use as input data.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from the input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.tl.pca function.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.tl.dendrogram">
<span class="sig-prename descclassname"><span class="pre">pyviper.tl.</span></span><span class="sig-name descname"><span class="pre">dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.tl.dendrogram" title="Permalink to this definition"></a></dt>
<dd><p>A wrapper for the scanpy function sc.tl.dendrogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Gene expression, protein activity or pathways stored in an anndata object.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>filter_by_feature_groups</em>) – The key in adata.uns where the dendrogram should be stored.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The layer to use as input data.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from the input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>**kwargs</strong> – Arguments to provide to the sc.tl.dendrogram function.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.tl.oncomatch">
<span class="sig-prename descclassname"><span class="pre">pyviper.tl.</span></span><span class="sig-name descname"><span class="pre">oncomatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pax_data_to_test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pax_data_for_cMRs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tcm_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both_ways</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_tail</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">om_max_NES_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">om_min_logp_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enrichment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aREA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'om'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.tl.oncomatch" title="Permalink to this definition"></a></dt>
<dd><p>The OncoMatch algorithm[1] assesses the overlap in differentially active MR
proteins between two sets of samples (e.g. to validate GEMMs as effective
models of human tumor samples). It does so by computing -log10 p-values for
each sample in pax_data_to_test of the MRs of each sample in pax_data_for_cMRs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pax_data_to_test</strong> – An anndata.AnnData or pd.DataFrame containing protein activity (NES),
where rows are observations/samples (e.g. cells or groups) and
columns are features (e.g. proteins or pathways).</p></li>
<li><p><strong>pax_data_for_cMRs</strong> – An anndata.AnnData or pd.DataFrame containing protein activity (NES),
where rows are observations/samples (e.g. cells or groups) and
columns are features (e.g. proteins or pathways).</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – Number of top MRs from each sample to use to compute regulators.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to also use the candidate MRs of pax_data_to_test to compute
NES for the samples in pax_data_for_cMRs, and then average.</p></li>
<li><p><strong>lower_tail</strong> (<em>default</em><em> (</em><em>True</em><em>)</em>) – If <cite>True</cite> (default), probabilities are P(X &lt;= x)
If <cite>False</cite>, probabilities are P(X &gt; x)</p></li>
<li><p><strong>(</strong><strong>default</strong> – The maximum NES scores before using a cutoff.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum logp value threshold, such that all logp values smaller than
this value are set to 0.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The method of compute enrichment. ‘aREA’ or ‘NaRnEA’</p></li>
<li><p><strong>(</strong><strong>default</strong> – The slot in pax_data_to_test.obsm to store the oncomatch results.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Instead of adding the OncoMatch DataFrame to pax_data_to_test.obsm,
return it directly.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>When copy is False, stores a pd.DataFrame objects of -log10 p-values with</em></p></li>
<li><p><em>shape (n_samples in pax_data_to_test, n_samples in pax_data_for_cMRs) in</em></p></li>
<li><p><em>pax_data_to_test.obsm[key_added]. When copy is True, a copy of the AnnData</em></p></li>
<li><p><em>is returned with these pd.DataFrames stored. When return_as_df is True,</em></p></li>
<li><p><em>the OncoMatch DataFrame alone is directly returned by the function.</em></p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>Alvarez, M. J. et al. A precision oncology approach to the pharmacological
targeting of mechanistic dependencies in neuroendocrine tumors. Nat Genet 50,
979–989, doi:10.1038/s41588-018-0138-4 (2018).</p>
<p>Alvarez, M. J. et al. Reply to ’H-STS, L-STS and KRJ-I are not authentic GEPNET
cell lines’. Nat Genet 51, 1427–1428, doi:10.1038/s41588-019-0509-5 (2019).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.tl.find_top_mrs">
<span class="sig-prename descclassname"><span class="pre">pyviper.tl.</span></span><span class="sig-name descname"><span class="pre">find_top_mrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca_slot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_column_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'stouffer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_feature_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_by_top_mrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.tl.find_top_mrs" title="Permalink to this definition"></a></dt>
<dd><p>Identify the top N master regulator proteins in a VIPER AnnData object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – An anndata object containing a distance object in adata.obsp.</p></li>
<li><p><strong>pca_slot</strong> – The slot in adata.obsm where a PCA is stored. Only required when method
is “spearman”.</p></li>
<li><p><strong>obs_column_name</strong> – The name of the column of observations in adata to use as clusters, or a
cluster vector corresponding to observations. Required when method is
“mwu” or “spearman”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>copy</em>) – The number of MRs to return</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to return both the top N and bottom N MRs (True) or just the
top N (False).</p></li>
<li><p><strong>(</strong><strong>default</strong> – The method used to compute a signature to identify the top candidate
master regulators (MRs). The options come from functions in pyviper.pp.
Choose between “stouffer”, “mwu”, or “spearman”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The name of the slot in the adata.var to store the output.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The selected regulators, such that all other regulators are filtered out
from the input data. If None, all regulators will be included. Regulator
sets must be from one of the following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – When False, a column is added to var with identified MRs labeled as
“True”, while all other proteins are labeled as “False”. When True, top
MRs are labeled N,N-1,N-2,…,1, bottom MRs are labeled -N,-N-1,-N-2,
…,-1, and all other proteins are labeled 0. Higher rank means greater
activity, while lower rank means less.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter var to only the top MRs in adata</p></li>
<li><p><strong>(</strong><strong>default</strong> – Returns a pd.DataFrame of the top MRs per cluster</p></li>
<li><p><strong>(</strong><strong>default</strong> – Determines whether a copy of the input AnnData is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Add a column to adata.var[key_added] or, when clusters given, adds multiple</em></p></li>
<li><p><em>columns (e.g. key_added_clust1name, key_added_clust2name, etc) to adata.var.</em></p></li>
<li><p><em>If copy, returns a new adata transformed by this function. If return_as_df,</em></p></li>
<li><p><em>returns a DataFrame.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.tl.path_enr">
<span class="sig-prename descclassname"><span class="pre">pyviper.tl.</span></span><span class="sig-name descname"><span class="pre">path_enr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway_interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enrichment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aREA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mvws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_as_anndata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transfer_obs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_input_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.tl.path_enr" title="Permalink to this definition"></a></dt>
<dd><p>Run the variation of VIPER that is specific to pathway enrichment analysis:
a single interactome and min_targets is set to 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy).</p></li>
<li><p><strong>pathway_interactome</strong> – An object of class Interactome or one of the following strings that
corresponds to msigdb regulons: “c2”, “c5”, “c6”, “c7”, “h”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>store_input_data</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – A method used to create a gene expression signature from gex_data.X. The
default of None is used when gex_data.X is already a gene expression
signature. Alternative inputs include “scale”, “rank”, “doublerank”,
“mad”, and “ttest”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The algorithm to use to calculate the enrichment. Choose betweeen
Analytical Ranked Enrichment Analysis (aREA) and Nonparametric
Analytical Rank-based Enrichment Analysis (NaRnEA) function. Default =’aREA’,
alternative = ‘NaRnEA’.</p></li>
<li><p><strong>(</strong><strong>default</strong> – (A) Number indicating either the exponent score for the metaViper weights.
These are only applicable when enrichment = ‘aREA’ and are not used when
enrichment = ‘NaRnEA’. Roughly, a lower number (e.g. 1) results in
networks being treated as a consensus network (useful for multiple
networks of the same celltype with the same epigenetics), while a higher
number (e.g. 10) results in networks being treated as separate (useful
for multiple networks of different celltypes with different epigenetics).
(B) The name of a column in gex_data that contains the manual assignments
of samples to networks using list position or network names.
(C) “auto”: assign samples to networks based on how well each
network allows for sample enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Number of cores to distribute sample batches into.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Maximum number of samples to process at once. Set to None to split all
samples across provided <cite>njobs</cite>.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Way of delivering output.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to transfer the observation metadata from the input anndata to
the output anndata. Thus, not applicable when output_as_anndata==False.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to store the input anndata in an unstructured data slot (.uns) of
the output anndata. Thus, not applicable when output_as_anndata==False.
If input anndata already contains ‘gex_data’ in .uns, the input will
assumed to be protein activity and will be stored in .uns as ‘pax_data’.
Otherwise, the data will be stored as ‘gex_data’ in .uns.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Returns an AnnData object containing the pathways. When store_input_data,</em></p></li>
<li><p><em>the input gex_data AnnData is stored within the dataframe.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.aREA.aREA_classic">
<span id="pyviper-area-area-classic"></span><h2>pyviper.aREA.aREA_classic<a class="headerlink" href="#module-pyviper.aREA.aREA_classic" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_classic.aREA_classic">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_classic.</span></span><span class="sig-name descname"><span class="pre">aREA_classic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_classic.aREA_classic" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the analytical ranked enrichment analysis (aREA)
function.</p>
<p>It is the basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – The interactome object.</p></li>
<li><p><strong>layer</strong> – The layer in the anndata object to use as the gene expression input
(default = None).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.aREA.aREA_meta">
<span id="pyviper-area-area-meta"></span><h2>pyviper.aREA.aREA_meta<a class="headerlink" href="#module-pyviper.aREA.aREA_meta" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_meta.aREA">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_meta.</span></span><span class="sig-name descname"><span class="pre">aREA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mvws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_meta.aREA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the Analytical Ranked Enrichment Analysis (aREA)[1]
function.</p>
<p>It is the original basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommended to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A consistent number of
targets allows regulators to have NES scores that are comparable to one
another. A regulator that has more targets than others will have “boosted”
NES scores, such that they cannot be compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – (A) Number indicating either the exponent score for the metaViper weights.
These are only applicable when enrichment = ‘area’ and are not used when
enrichment = ‘narnea’. Roughly, a lower number (e.g. 1) results in
networks being treated as a consensus network (useful for multiple
networks of the same celltype with the same epigenetics), while a higher
number (e.g. 10) results in networks being treated as separate (useful
for multiple networks of different celltypes with different epigenetics).
(B) The name of a column in gex_data that contains the manual assignments
of samples to networks using list position or network names.
(C) “auto”: assign samples to networks based on how well each
network allows for sample enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>[1] Alvarez, M. J., Shen, Y., Giorgi, F. M., Lachmann, A., Ding, B. B., Ye,</em></p></li>
<li><p><em>B. H., &amp; Califano, A. (2016). Functional characterization of somatic</em></p></li>
<li><p><em>mutations in cancer using network-based inference of protein activity.</em></p></li>
<li><p><em>Nature genetics, 48(8), 838-847.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.NaRnEA.NaRnEA_classic">
<span id="pyviper-narnea-narnea-classic"></span><h2>pyviper.NaRnEA.NaRnEA_classic<a class="headerlink" href="#module-pyviper.NaRnEA.NaRnEA_classic" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.NaRnEA.NaRnEA_classic.NaRnEA_classic">
<span class="sig-prename descclassname"><span class="pre">pyviper.NaRnEA.NaRnEA_classic.</span></span><span class="sig-name descname"><span class="pre">NaRnEA_classic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.NaRnEA.NaRnEA_classic.NaRnEA_classic" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-pyviper.NaRnEA.NaRnEA_meta">
<span id="pyviper-narnea-narnea-meta"></span><h2>pyviper.NaRnEA.NaRnEA_meta<a class="headerlink" href="#module-pyviper.NaRnEA.NaRnEA_meta" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.NaRnEA.NaRnEA_meta.NaRnEA">
<span class="sig-prename descclassname"><span class="pre">pyviper.NaRnEA.NaRnEA_meta.</span></span><span class="sig-name descname"><span class="pre">NaRnEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.NaRnEA.NaRnEA_meta.NaRnEA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores and proportional
enrichment scores from gene expression data using the Nonparametric
Analytical Rank-based Enrichment Analysis (NaRnEA)[1] function.</p>
<p>NaRnEA is an updated basis for the VIPER (Virtual Inference of
Protein-activity by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommend to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A regulator that has more
targets than others will have “boosted” NES scores, such that they cannot be
compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A dictionary containing <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code> containing NES values</p></li>
<li><p><strong>(key</strong> (<em>‘nes’) and PES values (key: ‘pes’).</em>)</p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>[1] Griffin, A. T., Vlahos, L. J., Chiuzan, C., &amp; Califano, A. (2023). NaRnEA</em></p></li>
<li><p><em>An Information Theoretic Framework for Gene Set Analysis. Entropy, 25(3), 542.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="pyVIPER (VIPER Analysis in Python for single-cell RNASeq)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Tutorial-1.html" class="btn btn-neutral float-right" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>