<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyviper package &mdash; pyviper 1.0.6 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" href="Tutorial-1.html" />
    <link rel="prev" title="pyVIPER (VIPER Analysis in Python for single-cell RNASeq)" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">pyviper package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pyviper-config">pyviper.config</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.interactome">pyviper.interactome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pyviper-pl">pyviper.pl</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pyviper-pp">pyviper.pp</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pyviper-tl">pyviper.tl</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.aREA.aREA_classic">pyviper.aREA.aREA_classic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.aREA.aREA_meta">pyviper.aREA.aREA_meta</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.NaRnEA.NaRnEA_classic">pyviper.NaRnEA.NaRnEA_classic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyviper.NaRnEA.NaRnEA_meta">pyviper.NaRnEA.NaRnEA_meta</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">pyviper package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pyviper.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="pyviper-package">
<h1>pyviper package<a class="headerlink" href="#pyviper-package" title="Permalink to this heading"></a></h1>
<div class="section" id="pyviper-config">
<h2>pyviper.config<a class="headerlink" href="#pyviper-config" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.config.set_regulators_filepath">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_filepath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.config.set_regulators_filepath" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to use a custom list of regulatory proteins instead of the
default ones within pyVIPER’s data folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group</strong> – A group of regulatory proteins of either: “tfs”, “cotfs”, “sig” or “surf”.</p></li>
<li><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p></li>
<li><p><strong>new_filepath</strong> – The new filepath that should be used to retrieve these sets of proteins.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyviper.config.set_regulators_species_to_use">
<span class="sig-prename descclassname"><span class="pre">pyviper.config.</span></span><span class="sig-name descname"><span class="pre">set_regulators_species_to_use</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.config.set_regulators_species_to_use" title="Permalink to this definition"></a></dt>
<dd><p>Allows the user to specify which species they are currently studying, so the
correct sets of regulatory proteins will be used during analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>species</strong> – The species to which the group of proteins belongs to: “human” or “mouse”.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.interactome">
<span id="pyviper-interactome"></span><h2>pyviper.interactome<a class="headerlink" href="#module-pyviper.interactome" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyviper.interactome.</span></span><span class="sig-name descname"><span class="pre">Interactome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.copy" title="Permalink to this definition"></a></dt>
<dd><p>Create a copy of this Interactome object.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An object of <a class="reference internal" href="#pyviper.interactome.Interactome" title="pyviper.interactome.Interactome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interactome</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.filter_regulators">
<span class="sig-name descname"><span class="pre">filter_regulators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regulators_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regulators_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.filter_regulators" title="Permalink to this definition"></a></dt>
<dd><p>Filter regulators by choosing by name or by group which ones you intend
to keep and which ones you intend to remove from this Interactome.</p>
<p>Note that the names of regulator that belong to the groups “tfs”, “cotfs”,
“sig” and “surf” will be sourced via the paths specified in pyviper.config.
To update these paths, use the pyviper.config.set_regulators_filepath
function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – This should be either:
(1) An array or list containing the names of specific regulators you
wish to keep in the network. When left as None, this parameter is
not used to filter.
(2) An array or list containing a group or groups of regulators that
you wish to keep in the network. These groups should be one of the
following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – This should be either:
(1) An array or list containing the names of specific regulators you
wish to remove from the network. When left as None, this parameter
is not used to filter.
(2) An array or list containing a group or groups of regulators that
you wish to remove from the network. These groups should be one of the
following: “tfs”, “cotfs”, “sig”, “surf”.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of regulators removed during filtering</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.filter_targets">
<span class="sig-name descname"><span class="pre">filter_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">targets_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targets_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.filter_targets" title="Permalink to this definition"></a></dt>
<dd><p>Filter targets by choosing by name which ones you intend to keep and
which ones you intend remove from this Interactome.</p>
<p>When working with an anndata object or a gene expression array, it is
highly recommended to filter the unPruned network before pruning. This
is to ensure the pruned network contains a consistent number of targets
per regulator regulator, all of which exist within gex_data. A regulator
that has more targets than others will have “boosted” NES scores, such
that they cannot be compared to those with fewer targets.
For example, with an anndata object named gex_data, one may is suggested
to do:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – An array containing the names of targets you wish to keep in the
network. When left as None, this parameter is not used to filter.</p></li>
<li><p><strong>(</strong><strong>default</strong> – An array containing the names of targets you wish to remove from the
network. When left as None, this parameter is not used to filter.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of targets removed during filtering</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_reg">
<span class="sig-name descname"><span class="pre">get_reg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regName</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_reg" title="Permalink to this definition"></a></dt>
<dd><p>Get the rows of the net_table where the regulator is regName.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regName</strong> – The name of a regulator in this Interactome.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_reg_names">
<span class="sig-name descname"><span class="pre">get_reg_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_reg_names" title="Permalink to this definition"></a></dt>
<dd><p>Get an array of all unique regulators in this Interactome.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An array of strings of <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.get_target_names">
<span class="sig-name descname"><span class="pre">get_target_names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.get_target_names" title="Permalink to this definition"></a></dt>
<dd><p>Get a set of the unique targets in this Interactome</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A 1D NumPy array.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.ic_mat">
<span class="sig-name descname"><span class="pre">ic_mat</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.ic_mat" title="Permalink to this definition"></a></dt>
<dd><p>Get the DataFrame of all the likelihood values. Targets are in the rows,
while Regulators are in the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.icp_vec">
<span class="sig-name descname"><span class="pre">icp_vec</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.icp_vec" title="Permalink to this definition"></a></dt>
<dd><p>Get the vector containing the proportion of the “Interaction Confidence”
(IC) score for each interaction in a network, relative to the maximum IC
score in the network. This vector is generated by taking each individual
regulon in the newtork and calculating the likelihood index proportion
to all interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An array of <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.integrate">
<span class="sig-name descname"><span class="pre">integrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_likelihoods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.integrate" title="Permalink to this definition"></a></dt>
<dd><p>Integrate this Interactome object with one or more other Interacome
objects to create a consensus network. In general, this should be done
when interactome objects have the same epigenetics (e.g. due to being
made from different datasets of same celltype). MetaVIPER should be used
instead when you have multiple interactomes with different epigenetics
(e.g. due to being made of data with different celltypes).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>network_list</strong> – A single object or a list of objects of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>normalize_likelihoods</em>) – An array containing weights for each network being integrated. The
first weight corresponds to this network, while the others correspond
to those in the network list in order. If None, equal weights are
used.</p></li>
<li><p><strong>(</strong><strong>default</strong> – An extra operation that can performed after the integration
operation where within each regulator, likelihood values are ranked
and scaled from 0 to 1.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.mor_mat">
<span class="sig-name descname"><span class="pre">mor_mat</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.mor_mat" title="Permalink to this definition"></a></dt>
<dd><p>Get the DataFrame of all the correlation values. Targets are in the rows,
while Regulators are in the columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.prune">
<span class="sig-name descname"><span class="pre">prune</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eliminate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.prune" title="Permalink to this definition"></a></dt>
<dd><p>Prune the Interactome by eliminating extra targets from regulators and,
with eliminate = True, remove regulators with too few targets from the
network. Note that by ensuring the pruned networks contains the same
number of targets for each regulator, NES scores are comparable. If one
regulator has more targest than another, than its NES score will be
“boosted” and they cannot be compared against each other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The maximum number of targets that each regulon is allowed.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulon is required.</p></li>
<li><p><strong>(</strong><strong>default</strong> – If eliminate = True, then any regulators with fewer targets than
max_targets will be removed from the network. In other words, after
pruning, all regulators will have exactly max_targets number of
targets. This essentially sets min_targets equal to max_targets
and ensures all NES scores are comparable with aREA.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Report the number of targets and regulators removed during pruning</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.save" title="Permalink to this definition"></a></dt>
<dd><p>Save the Interactome object to one’s disk. If saved as “csv” or
“tsv”, just the interactome.net_table will be saved. If saved as
“pkl”, the whole interactome object will be saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> – A filepath to one’s disk to store the Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>output_type</em>) – If None, the output_type will be inferred from the file_path.
Otherwise, specify “csv”, “tsv” or “pkl”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.size" title="Permalink to this definition"></a></dt>
<dd><p>Get the the number of regulators in this Interactome.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>An int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.translate_regulators">
<span class="sig-name descname"><span class="pre">translate_regulators</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">desired_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.translate_regulators" title="Permalink to this definition"></a></dt>
<dd><p>Translate the regulators of the Interactome. The current name format of
the regulators should be one of the following:</p>
<blockquote>
<div><p>mouse_symbol, mouse_ensembl, mouse_entrez, human_symbol, human_ensembl or human_entrez</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>desired_format</strong> – Desired format can be one of four strings: “mouse_symbol”,
“mouse_ensembl”, “mouse_entrez”, “human_symbol”, “human_ensembl” or “human_entrez”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – Report the number of regulators successfully and unsucessfully translated</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyviper.interactome.Interactome.translate_targets">
<span class="sig-name descname"><span class="pre">translate_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">desired_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.interactome.Interactome.translate_targets" title="Permalink to this definition"></a></dt>
<dd><p>Translate the targets of the Interactome.  The current name format of
the targets should be one of the following:</p>
<blockquote>
<div><p>mouse_symbol, mouse_ensembl, mouse_entrez, human_symbol, human_ensembl or human_entrez</p>
</div></blockquote>
<p>It is recommended to do this before pruning to ensure a consistent number
of targets because if targets do not have a translation, they will be
deleted, resulting in different numbers of targets in a pruned
interactome that once had consistent number of targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>desired_format</strong> – Desired format can be one of four strings: “mouse_symbol”,
“mouse_ensembl”, “mouse_entrez”, “human_symbol”, “human_ensembl” or “human_entrez”.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – Report the number of targets successfully and unsucessfully translated</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="pyviper-pl">
<h2>pyviper.pl<a class="headerlink" href="#pyviper-pl" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="pyviper-pp">
<h2>pyviper.pp<a class="headerlink" href="#pyviper-pp" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="pyviper-tl">
<h2>pyviper.tl<a class="headerlink" href="#pyviper-tl" title="Permalink to this heading"></a></h2>
</div>
<div class="section" id="module-pyviper.aREA.aREA_classic">
<span id="pyviper-area-area-classic"></span><h2>pyviper.aREA.aREA_classic<a class="headerlink" href="#module-pyviper.aREA.aREA_classic" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_classic.aREA_classic">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_classic.</span></span><span class="sig-name descname"><span class="pre">aREA_classic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_classic.aREA_classic" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the analytical ranked enrichment analysis (aREA)
function.</p>
<p>It is the basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – The interactome object.</p></li>
<li><p><strong>layer</strong> – The layer in the anndata object to use as the gene expression input
(default = None).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.aREA.aREA_meta">
<span id="pyviper-area-area-meta"></span><h2>pyviper.aREA.aREA_meta<a class="headerlink" href="#module-pyviper.aREA.aREA_meta" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.aREA.aREA_meta.aREA">
<span class="sig-prename descclassname"><span class="pre">pyviper.aREA.aREA_meta.</span></span><span class="sig-name descname"><span class="pre">aREA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mvws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.aREA.aREA_meta.aREA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores from gene
expression data using the Analytical Ranked Enrichment Analysis (aREA)[1]
function.</p>
<p>It is the original basis of the VIPER (Virtual Inference of Protein-activity
by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommended to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A consistent number of
targets allows regulators to have NES scores that are comparable to one
another. A regulator that has more targets than others will have “boosted”
NES scores, such that they cannot be compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – (A) Number indicating either the exponent score for the metaViper weights.
These are only applicable when enrichment = ‘area’ and are not used when
enrichment = ‘narnea’. Roughly, a lower number (e.g. 1) results in
networks being treated as a consensus network (useful for multiple
networks of the same celltype with the same epigenetics), while a higher
number (e.g. 10) results in networks being treated as separate (useful
for multiple networks of different celltypes with different epigenetics).
(B) The name of a column in gex_data that contains the manual assignments
of samples to networks using list position or network names.
(C) “auto”: assign samples to networks based on how well each
network allows for sample enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A dataframe of <code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code> containing NES values.</p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>[1] Alvarez, M. J., Shen, Y., Giorgi, F. M., Lachmann, A., Ding, B. B., Ye,</em></p></li>
<li><p><em>B. H., &amp; Califano, A. (2016). Functional characterization of somatic</em></p></li>
<li><p><em>mutations in cancer using network-based inference of protein activity.</em></p></li>
<li><p><em>Nature genetics, 48(8), 838-847.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pyviper.NaRnEA.NaRnEA_classic">
<span id="pyviper-narnea-narnea-classic"></span><h2>pyviper.NaRnEA.NaRnEA_classic<a class="headerlink" href="#module-pyviper.NaRnEA.NaRnEA_classic" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.NaRnEA.NaRnEA_classic.NaRnEA_classic">
<span class="sig-prename descclassname"><span class="pre">pyviper.NaRnEA.NaRnEA_classic.</span></span><span class="sig-name descname"><span class="pre">NaRnEA_classic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.NaRnEA.NaRnEA_classic.NaRnEA_classic" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-pyviper.NaRnEA.NaRnEA_meta">
<span id="pyviper-narnea-narnea-meta"></span><h2>pyviper.NaRnEA.NaRnEA_meta<a class="headerlink" href="#module-pyviper.NaRnEA.NaRnEA_meta" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyviper.NaRnEA.NaRnEA_meta.NaRnEA">
<span class="sig-prename descclassname"><span class="pre">pyviper.NaRnEA.NaRnEA_meta.</span></span><span class="sig-name descname"><span class="pre">NaRnEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gex_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eset_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyviper.NaRnEA.NaRnEA_meta.NaRnEA" title="Permalink to this definition"></a></dt>
<dd><p>Allows the individual to infer normalized enrichment scores and proportional
enrichment scores from gene expression data using the Nonparametric
Analytical Rank-based Enrichment Analysis (NaRnEA)[1] function.</p>
<p>NaRnEA is an updated basis for the VIPER (Virtual Inference of
Protein-activity by Enriched Regulon analysis) algorithm.</p>
<p>The Interactome object must not contain any targets that are not in the
features of gex_data. This can be accomplished by running:</p>
<blockquote>
<div><p>interactome.filter_targets(gex_data.var_names)</p>
</div></blockquote>
<p>It is highly recommend to do this on the unPruned network and then prune to
ensure the pruned network contains a consistent number of targets per
regulator, all of which exist within gex_data. A regulator that has more
targets than others will have “boosted” NES scores, such that they cannot be
compared to those with fewer targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gex_data</strong> – Gene expression stored in an anndata object (e.g. from Scanpy) or in a
pd.DataFrame.</p></li>
<li><p><strong>interactome</strong> – An object of class Interactome.</p></li>
<li><p><strong>(</strong><strong>default</strong> (<em>verbose</em>) – The layer in the anndata object to use as the gene expression input.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether to filter out genes not present in the interactome (True) or to
keep this biological context (False). This will affect gene rankings.</p></li>
<li><p><strong>(</strong><strong>default</strong> – The minimum number of targets that each regulator in the interactome
should contain. Regulators that contain fewer targets than this minimum
will be culled from the network (via the Interactome.cull method). The
reason users may choose to use this threshold is because adequate
targets are needed to accurately predict enrichment.</p></li>
<li><p><strong>(</strong><strong>default</strong> – Whether extended output about the progress of the algorithm should be
given.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A dictionary containing <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code> containing NES values</p></li>
<li><p><strong>(key</strong> (<em>‘nes’) and PES values (key: ‘pes’).</em>)</p></li>
<li><p><em>Citations</em></p></li>
<li><p><em>——-</em></p></li>
<li><p><em>[1] Griffin, A. T., Vlahos, L. J., Chiuzan, C., &amp; Califano, A. (2023). NaRnEA</em></p></li>
<li><p><em>An Information Theoretic Framework for Gene Set Analysis. Entropy, 25(3), 542.</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="pyVIPER (VIPER Analysis in Python for single-cell RNASeq)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Tutorial-1.html" class="btn btn-neutral float-right" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>