<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach &mdash; pyviper 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis" href="Tutorial-3.html" />
    <link rel="prev" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" href="Tutorial-1.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Install-PyVIPER">Install PyVIPER</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Import-modules">Import modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">Step 1. Load a gene expression matrix and associated metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-2.-Preprocess-and-generate-a-gene-expression-signature-at-the-single-cell-level">Step 2. Preprocess and generate a gene expression signature at the single-cell level</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-3.-Load-multiple-ARACNe-inferred-gene-regulatory-networks">Step 3. Load multiple ARACNe-inferred gene regulatory networks</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Convert-the-gene-expression-signature-into-a-protein-activity-matrix-using-pyviper-(metaVIPER-approach)">Convert the gene expression signature into a protein activity matrix using pyviper (metaVIPER approach)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Step-4.-Analyze-single-cells-at-the-Protein-Activity-level">Step 4. Analyze single-cells at the Protein Activity level</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Pathway-enrichment-analysis">Pathway enrichment analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Key-takeaways">Key takeaways</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Tutorial-2.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="Tutorial-2---Inferring-Protein-Activity-from-scRNA-seq-data-from-multiple-cell-populations-with-the-meta-VIPER-approach">
<h1>Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach<a class="headerlink" href="#Tutorial-2---Inferring-Protein-Activity-from-scRNA-seq-data-from-multiple-cell-populations-with-the-meta-VIPER-approach" title="Permalink to this heading"></a></h1>
<p>This tutorial explores the use of metaVIPER to assess protein activity by integrative analysis using multiple gene regulatory networks. metaVIPER is particularly useful in real-world scenarios to infer protein activity in a tissue-independent manner by using multiple, non-tissue-matched interactomes, assuming that transcriptional targets of each protein can be recapitulated by one or more available interactomes. The <code class="docutils literal notranslate"><span class="pre">viper</span></code> function allows running metaVIPER using both
<a class="reference external" href="https://www.nature.com/articles/ng.3593">aREA</a> and <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">NaRnEA</a> enrichment methods. For additional details, please refer to <a class="reference external" href="https://www.nature.com/articles/s41467-018-03843-3">Ding et al., 2016</a>. After a brief description of the installation procedure and the modules needed, this notebook is organized in the following sections.</p>
<p><strong>Table of Contents</strong></p>
<div class="line-block">
<div class="line">Step 1. Load a gene expression matrix and associated metadata</div>
<div class="line">Step 2. Preprocess and generate a gene expression signature at the single-cell level</div>
<div class="line">Step 3. Load multiple ARACNe-inferred gene regulatory networks</div>
<div class="line">Step 4. Analyze single-cells at the Protein Activity level</div>
<div class="line">Step 5. Pathway enrichment analysis</div>
<div class="line">Key Takeaways</div>
</div>
<div class="section" id="Install-PyVIPER">
<h2>Install PyVIPER<a class="headerlink" href="#Install-PyVIPER" title="Permalink to this heading"></a></h2>
<p>Install <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> from PyPI using pip. Alternatively, refer to the README in the current GitHub to install from the local directory.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># !pip install pyviper</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Import-modules">
<h2>Import modules<a class="headerlink" href="#Import-modules" title="Permalink to this heading"></a></h2>
<p>Load <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> and additional modules required used in this tutorial.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyviper</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>


<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;.*The &#39;nopython&#39; keyword.*&quot;</span><span class="p">)</span> <span class="c1"># for jit decorator issue with sc.pp.neighbors (09/30/2023)</span>
</pre></div>
</div>
</div>
<p>Setup the path to the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_location</span> <span class="o">=</span> <span class="s2">&quot;https://zenodo.org/records/10059791/files/&quot;</span>
</pre></div>
</div>
</div>
<p>We will import the data directly from the above link. Optionally, users are allowed to download the files from <a class="reference external" href="https://zenodo.org/records/10059791">here</a> and to load them from their local computer. If the second option is preferred, the following lines can be uncommented and run to download the data on Unix systems and set the working directory to the path where the data will be stored.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!mkdir Tutorial_2_data_dir</span>
<span class="c1">#%cd Tutorial_2_data_dir</span>
<span class="c1">#!curl -O &quot;{data_location}Tutorial_2_counts_mixed_4632.tsv.gz&quot; -O &quot;{data_location}Tutorial_2_metadata_mixed_4632.tsv.gz&quot;</span>
<span class="c1">#!curl -O &quot;{data_location}B-cell-net.tsv&quot; -O &quot;{data_location}ductal-2-net.tsv&quot; -O &quot;{data_location}fibroblast-net.tsv&quot; -O &quot;{data_location}stellate-net.tsv&quot;</span>
<span class="c1">#data_location = !pwd</span>
<span class="c1">#data_location = data_location[0]+&quot;/&quot; # new value for data if working on data downloaded locally</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">
<h2>Step 1. Load a gene expression matrix and associated metadata<a class="headerlink" href="#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata" title="Permalink to this heading"></a></h2>
<p>Load the gene expression matrix (UMIs) and store it into an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">AnnData</a> object to enable interoperability with <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/#">scanpy</a>. Cells used in this tutorial were sampled from scRNA-seq data published in <a class="reference external" href="https://www.nature.com/articles/s41422-019-0195-y">Peng et al., 2019</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gExpr_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;Tutorial_2_counts_mixed_4632.tsv.gz&quot;</span> <span class="c1"># path to gene expression matrix (UMI counts)</span>
<span class="n">adata_gExpr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gExpr_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># read from remote</span>
<span class="n">adata_gExpr</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span> <span class="c1"># convert to AnnData object</span>
</pre></div>
</div>
</div>
<p>Display matrix dimensions (cells x genes)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4632 × 20197
    obs: &#39;Cell_Type&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;
    var: &#39;n_cells&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;log1p&#39;
</pre></div></div>
</div>
<p>Load cell-associated metadata.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;Tutorial_2_metadata_mixed_4632.tsv.gz&quot;</span> <span class="c1"># path to cells metadata</span>
<span class="n">cells_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># load it</span>
</pre></div>
</div>
</div>
<p>Store the metadata in the <code class="docutils literal notranslate"><span class="pre">adata_gExpr</span></code> object as observation annotation.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">cells_metadata</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span><span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># store cell-specific metadata as annotation observation</span>
</pre></div>
</div>
</div>
<p>Display the observation annotation from the AnnData object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell_Type</th>
      <th>n_genes</th>
      <th>n_genes_by_counts</th>
      <th>total_counts</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AACCATGCACAACTGT</th>
      <td>Ductal cell type 2</td>
      <td>1218</td>
      <td>1218</td>
      <td>2395.170410</td>
    </tr>
    <tr>
      <th>T1_AAGACCTAGTCATGCT</th>
      <td>Ductal cell type 2</td>
      <td>3149</td>
      <td>3146</td>
      <td>3325.380615</td>
    </tr>
    <tr>
      <th>T1_ACATACGAGACTCGGA</th>
      <td>Ductal cell type 2</td>
      <td>1044</td>
      <td>1044</td>
      <td>2051.984863</td>
    </tr>
    <tr>
      <th>T1_ACCAGTATCTTGCAAG</th>
      <td>Ductal cell type 2</td>
      <td>4039</td>
      <td>4037</td>
      <td>3812.823730</td>
    </tr>
    <tr>
      <th>T1_ACGATGTTCACGAAGG</th>
      <td>Ductal cell type 2</td>
      <td>3377</td>
      <td>3375</td>
      <td>3621.977051</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The observation annotations include the annotated cell type for each single-cell. List the cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Cell_Type&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;n&#39;</span><span class="p">)</span> <span class="c1"># show cell types and number of cells for each type in AnnData</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell_Type</th>
      <th>n</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>B cell</td>
      <td>1000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Ductal cell type 2</td>
      <td>1455</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Fibroblast cell</td>
      <td>1277</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Stellate cell</td>
      <td>900</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Data contain 4 cells types: B cell, (malignant) ductal cells of type 2, fibroblasts, and stellate cells (myofibroblast-like cells).</p>
</div>
<div class="section" id="Step-2.-Preprocess-and-generate-a-gene-expression-signature-at-the-single-cell-level">
<h2>Step 2. Preprocess and generate a gene expression signature at the single-cell level<a class="headerlink" href="#Step-2.-Preprocess-and-generate-a-gene-expression-signature-at-the-single-cell-level" title="Permalink to this heading"></a></h2>
<p>The UMI matrix can be processed using the standard <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> preprocessing workflow. For a more detailed explanation of quality control steps, refer to the preprocessing tutorials by <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">scanpy</a> or <a class="reference external" href="https://satijalab.org/seurat/articles/pbmc3k_tutorial">Seurat</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span> <span class="c1"># filter out cells with &lt;200 genes expressed</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># filter out genes that are detected in &lt;3 cells</span>
</pre></div>
</div>
</div>
<p>Display matrix dimensions post-filtering (cells x genes)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4632 × 20197
    obs: &#39;Cell_Type&#39;, &#39;n_genes&#39;
    var: &#39;n_cells&#39;
</pre></div></div>
</div>
<p>Normalize the gene expression matrix by the total number of UMIs to have 10,000 UMIs per cells and log-transform the normalized counts to make each gene expression comparable across cells. Then, store the log-normalized counts in the <code class="docutils literal notranslate"><span class="pre">.raw</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span> <span class="c1"># normalize counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span> <span class="c1"># log-transform the counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># compute some statistics, e.g. total_counts, n_cells_by_counts etc</span>

<span class="n">adata_gExpr</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata_gExpr</span>   <span class="c1"># store log-normalized counts to .raw Attribute for potential future use</span>
</pre></div>
</div>
</div>
<p>Generate a gene expression signature on a single-cell basis by scaling each gene to unit variance. This scaled matrix will be used as an input to <code class="docutils literal notranslate"><span class="pre">viper</span></code> to compute the protein activity matrix. Several approaches exist for the purpose. These include the calculation of a signature with respect to an external reference or the use of models such as <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">SCTransform</a>. We will employ the simplest approach for demonstrative
purposes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># sc.pp.regress_out(adata_gExpr, &#39;total_counts&#39;) # regress out the effect of the total number of UMIs</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="c1"># standardize gene expression and clip max values to 10</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4632 × 20197
    obs: &#39;Cell_Type&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;
    var: &#39;n_cells&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;log1p&#39;
</pre></div></div>
</div>
<p>Display a chunk of the scaled (i.e. the signature) matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AP006222.2</th>
      <th>RP11-206L10.3</th>
      <th>RP11-206L10.2</th>
      <th>RP11-206L10.9</th>
      <th>LINC00115</th>
      <th>FAM41C</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AAGACCTAGTCATGCT</th>
      <td>-0.445240</td>
      <td>-0.125835</td>
      <td>-0.059886</td>
      <td>-0.114998</td>
      <td>-0.220759</td>
      <td>-0.107077</td>
    </tr>
    <tr>
      <th>T1_ACATACGAGACTCGGA</th>
      <td>-0.282344</td>
      <td>-0.082131</td>
      <td>-0.067132</td>
      <td>-0.040628</td>
      <td>-0.145249</td>
      <td>-0.152253</td>
    </tr>
    <tr>
      <th>T1_ACCAGTATCTTGCAAG</th>
      <td>-0.507595</td>
      <td>-0.142565</td>
      <td>-0.057112</td>
      <td>-0.143467</td>
      <td>-0.249664</td>
      <td>-0.089785</td>
    </tr>
    <tr>
      <th>T1_ACGATGTTCACGAAGG</th>
      <td>-0.483181</td>
      <td>-0.136015</td>
      <td>-0.058198</td>
      <td>-0.132320</td>
      <td>-0.238347</td>
      <td>-0.096555</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Step-3.-Load-multiple-ARACNe-inferred-gene-regulatory-networks">
<h2>Step 3. Load multiple ARACNe-inferred gene regulatory networks<a class="headerlink" href="#Step-3.-Load-multiple-ARACNe-inferred-gene-regulatory-networks" title="Permalink to this heading"></a></h2>
<p>Load 4 lineage-specific <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">ARACNe3</a>-inferred gene regulatory networks. These networks were generated from B cells, malignant ductal, fibroblasts and stellate cells, respectively. Refer to the <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">ARACNe3</a> and <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/27153652/">ARACNe-AP</a> manuscripts for additional information. The time required is approximately 1 s per network (when loading from local) and 40 s per network (when
loading from the remote url).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">b_net_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;B-cell-net.tsv&quot;</span> <span class="c1"># path to ARACNe network for B cells</span>
<span class="n">d2_net_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;ductal-2-net.tsv&quot;</span> <span class="c1"># path to ARACNe network for ductal 2</span>
<span class="n">f_net_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;fibroblast-net.tsv&quot;</span> <span class="c1"># path to ARACNe network for fibroblasts</span>
<span class="n">s_net_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;stellate-net.tsv&quot;</span> <span class="c1"># path to ARACNe network for stellate cells</span>

<span class="n">bcell_net</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">Interactome</span><span class="p">(</span><span class="s1">&#39;b_cell&#39;</span><span class="p">,</span><span class="n">net_table</span><span class="o">=</span><span class="n">b_net_path</span><span class="p">)</span> <span class="c1"># load interactome for B cells</span>
<span class="n">ductal_2_net</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">Interactome</span><span class="p">(</span><span class="s1">&#39;ductal_2&#39;</span><span class="p">,</span><span class="n">net_table</span><span class="o">=</span><span class="n">d2_net_path</span><span class="p">)</span> <span class="c1"># load interactome for ductal cell type 2</span>
<span class="n">fibroblast_net</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">Interactome</span><span class="p">(</span><span class="s1">&#39;fibroblast&#39;</span><span class="p">,</span><span class="n">net_table</span><span class="o">=</span><span class="n">f_net_path</span><span class="p">)</span> <span class="c1"># load interactome for fibroblasts</span>
<span class="n">stellate_net</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">Interactome</span><span class="p">(</span><span class="s1">&#39;stellate&#39;</span><span class="p">,</span><span class="n">net_table</span><span class="o">=</span><span class="n">s_net_path</span><span class="p">)</span> <span class="c1"># load interactome for stellate cells</span>
</pre></div>
</div>
</div>
<p>Each regulatory network loaded as an instance of class <code class="docutils literal notranslate"><span class="pre">Interactome</span></code>:</p>
<p>Due to the similarity between regulatory networks of fibroblasts and stellate cells, we integrate their corresponding interactomes, thus generating a consensus network.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fibroblast_net</span><span class="o">.</span><span class="n">integrate</span><span class="p">(</span><span class="n">stellate_net</span><span class="p">,</span> <span class="n">normalize_likelihoods</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Optionally, save the integrated interactome as a .tsv file if needed for later use, by uncommenting the following line. Possible output formats are: “csv”, “tsv” and “pkl”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#fibroblast_net.save(&quot;fibroblast_integrated.tsv&quot;, output_type=&quot;tsv&quot;)</span>
</pre></div>
</div>
</div>
<p>Filter out targets in the <code class="docutils literal notranslate"><span class="pre">Interactomes</span></code> that are not present in the gene expression matrix</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bcell_net</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">ductal_2_net</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
<span class="n">fibroblast_net</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Removed 38885 targets.
Removed 11073 targets.
Removed 84211 targets.
</pre></div></div>
</div>
<p>Prune each regulon to have exactly 100 targets (this step can be omitted when setting <code class="docutils literal notranslate"><span class="pre">enrichment=narnea</span></code> in <code class="docutils literal notranslate"><span class="pre">pyviper</span></code>)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bcell_net</span><span class="o">.</span><span class="n">prune</span><span class="p">(</span><span class="n">max_targets</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">eliminate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ductal_2_net</span><span class="o">.</span><span class="n">prune</span><span class="p">(</span><span class="n">max_targets</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">eliminate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">fibroblast_net</span><span class="o">.</span><span class="n">prune</span><span class="p">(</span><span class="n">max_targets</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">eliminate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Removed 2837239 targets.
Removed 283 regulators.
Removed 2118839 targets.
Removed 10 regulators.
Removed 6427448 targets.
Removed 64 regulators.
</pre></div></div>
</div>
<div class="section" id="Convert-the-gene-expression-signature-into-a-protein-activity-matrix-using-pyviper-(metaVIPER-approach)">
<h3>Convert the gene expression signature into a protein activity matrix using pyviper (metaVIPER approach)<a class="headerlink" href="#Convert-the-gene-expression-signature-into-a-protein-activity-matrix-using-pyviper-(metaVIPER-approach)" title="Permalink to this heading"></a></h3>
<p>Convert previously computed gene expression signature into protein activity by using multiple interactomes simultaneousely via the metaVIPER approach. Use <code class="docutils literal notranslate"><span class="pre">enrichment=&quot;narnea&quot;</span></code> with 1 core and store the result in an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object (default). See Tutorial 1 for details on the available modalities to compute enrichment.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_PA</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">gex_data</span><span class="o">=</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="c1"># gene expression signature</span>
                            <span class="n">interactome</span><span class="o">=</span><span class="p">[</span><span class="n">bcell_net</span><span class="p">,</span> <span class="n">ductal_2_net</span><span class="p">,</span> <span class="n">fibroblast_net</span><span class="p">],</span> <span class="c1"># list of interactomes</span>
                            <span class="n">enrichment</span> <span class="o">=</span> <span class="s2">&quot;narnea&quot;</span><span class="p">,</span>
                            <span class="n">njobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="c1"># 3 cores</span>
                            <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Preparing the association scores
Computing regulons enrichment with NaRnEa
0/3 networks complete.
reordering genes
Calculating DES...
Calculating UES...
Calculating NES...
Calculating PES...
1/3 networks complete.
reordering genes
Calculating DES...
Calculating UES...
Calculating NES...
Calculating PES...
2/3 networks complete.
reordering genes
Calculating DES...
Calculating UES...
Calculating NES...
Calculating PES...
3/3 networks complete.
Integrating results
</pre></div></div>
</div>
<p>The output <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object stores a matrix collecting the Normalized Enrichment Scores (NES) and a matrix of Proportional Enrichment Scores (PES) with cells on rows and regulatory proteins on columns. See Tutorial 1 and <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/36981431/">Griffin et al., 2022</a> for more details .</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_PA</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4632 × 50
    obs: &#39;Cell_Type&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;
    uns: &#39;gex_data&#39;, &#39;pax_data&#39;
    layers: &#39;pes&#39;
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Step-4.-Analyze-single-cells-at-the-Protein-Activity-level">
<h2>Step 4. Analyze single-cells at the Protein Activity level<a class="headerlink" href="#Step-4.-Analyze-single-cells-at-the-Protein-Activity-level" title="Permalink to this heading"></a></h2>
<p>We present some instructive analyses at the protein activity level. Start by running Principal Components Analysis (PCA) on the PES matrix, to reduce the dimensionality of the protein activity space. Use the PCA function avaiable through <code class="docutils literal notranslate"><span class="pre">pyviper.tl</span></code>, a module that provides several wrappers to <code class="docutils literal notranslate"><span class="pre">scanpy.tl</span></code> to simplify the transformation of protein activity data. In contrast to Tutorial 1, all recovered transcription factors, cofactors, signaling and surface proteins will be employed in the
analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pyviper</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;pes&quot;</span><span class="p">,</span> <span class="n">filter_by_feature_groups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(4632, 6659)
</pre></div></div>
</div>
<p>Compute the neighbors graph of cells using the PCA representation of the protein activity matrix. For sake of simplicity, we set 10 nearest neightbors and 30 principal components. Typically, these parameters need tuning.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Generate a <a class="reference external" href="https://arxiv.org/abs/1802.03426">UMAP</a> embedding of the previously computed graph.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Display the 2-dimensional UMAP embedding and color by cell type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">pyviper</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set3&quot;</span><span class="p">,</span> <span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-2_55_0.png" class="no-scaled-link" src="_images/Tutorial-2_55_0.png" style="width: 478px; height: 333px;" />
</div>
</div>
<p>As an example, show the activity of selected markers for of ductal cells and fibroblasts in a dotplot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protein_markers_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;B cell&#39;</span><span class="p">:</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Ductal cell type 2&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;KRT19&#39;</span><span class="p">,</span> <span class="s1">&#39;TSPAN8&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Fibroblast cell&#39;</span><span class="p">:</span> <span class="s1">&#39;VIM&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Stellate cell&#39;</span><span class="p">:</span> <span class="s1">&#39;GEM&#39;</span>
<span class="p">}</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">pyviper</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">var_names</span><span class="o">=</span><span class="n">protein_markers_dict</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;pes&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;Cell_Type&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;RdBu_r&quot;</span><span class="p">,</span> <span class="n">vcenter</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-2_57_0.png" class="no-scaled-link" src="_images/Tutorial-2_57_0.png" style="width: 426px; height: 360px;" />
</div>
</div>
<p>Identify differentially active proteins on a cell type basis using the Wilcoxon Rank sum test and display a heatmap.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># reduce the verbosity</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;Cell_Type&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;pes&quot;</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>  <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilcoxon&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_heatmap</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">25</span><span class="p">,</span> <span class="n">vcenter</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ranking genes
    finished (0:00:03)
WARNING: dendrogram data not found (using key=dendrogram_Cell_Type). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.
    using &#39;X_pca&#39; with n_pcs = 50
Storing dendrogram info using `.uns[&#39;dendrogram_Cell_Type&#39;]`
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/av2729/opt/anaconda3/envs/single-cell-env-scvi/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: RuntimeWarning: invalid value encountered in log2
  self.stats[group_name, &#39;logfoldchanges&#39;] = np.log2(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-2_59_2.png" class="no-scaled-link" src="_images/Tutorial-2_59_2.png" style="width: 1248px; height: 712px;" />
</div>
</div>
</div>
<div class="section" id="Pathway-enrichment-analysis">
<h2>Pathway enrichment analysis<a class="headerlink" href="#Pathway-enrichment-analysis" title="Permalink to this heading"></a></h2>
<p>Perform an enrichment analysis to identify biological functions associated to malignant cells by leveraging functional gene sets from the <a class="reference external" href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a> collection. Through its <code class="docutils literal notranslate"><span class="pre">load</span></code> module,<code class="docutils literal notranslate"><span class="pre">pyviper</span></code> allows the user to choose between several MSigDB collections, including curated gene sets from online pathway databases (‘c2’), ontology (‘c5’) and oncogenic signature (‘c6’) gene sets and immunologic signature gene sets (‘c7’).</p>
<p>Load the ‘hallmark gene sets’ as an object of class <code class="docutils literal notranslate"><span class="pre">Interactome</span></code> and retain only the targets in common with the protein activity matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hallmarks_MSigDB</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">load</span><span class="o">.</span><span class="n">msigdb_regulon</span><span class="p">(</span><span class="s1">&#39;h&#39;</span><span class="p">)</span> <span class="c1"># load hallmark of cancer MSigDB as an interactome</span>
<span class="n">hallmarks_MSigDB</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">adata_PA</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Removed 3587 targets.
</pre></div></div>
</div>
<p>Perform the enrichment analysis using <code class="docutils literal notranslate"><span class="pre">viper</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_PA</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">adata_PA</span><span class="p">,</span> <span class="n">interactome</span><span class="o">=</span><span class="n">hallmarks_MSigDB</span><span class="p">,</span> <span class="n">enrichment</span><span class="o">=</span><span class="s2">&quot;narnea&quot;</span><span class="p">,</span> <span class="n">min_targets</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="s2">&quot;False&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Preparing the association scores
Computing regulons enrichment with NaRnEa
reordering genes
Calculating DES...
Calculating UES...
Calculating NES...
Calculating PES...
</pre></div></div>
</div>
<p>Focus on malignant ductal cells and compute the mean NES for each gene set.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_DC2</span> <span class="o">=</span> <span class="n">adata_PA</span><span class="p">[</span><span class="n">adata_PA</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Cell_Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Ductal cell type 2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">Integrated_hallmarks_NES</span> <span class="o">=</span> <span class="n">adata_DC2</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1">#Integrated_hallmarks_NES = pyviper.tl.stouffer_clusters_df(adata_DC2.to_df(), cluster_vector=[&quot;Ductal cell type 2&quot;] * adata_DC2.shape[0])</span>
</pre></div>
</div>
</div>
<p>Convert NES for each hallmark into p values and adjust for multiple hypothesis testing with the <a class="reference external" href="https://www.jstor.org/stable/2346101">Benjamini-Hochberg</a> procedure.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Integrated_hallmarks_pvals</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">nes_to_pval_df</span><span class="p">(</span><span class="n">Integrated_hallmarks_NES</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Compute the -log10(P value) and display the top 5 enriched pathways.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Integrated_hallmarks_mlog10</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">Integrated_hallmarks_pvals</span><span class="p">)</span> <span class="c1"># convert to -log10(P adjust)</span>

<span class="n">top_pathways</span> <span class="o">=</span> <span class="n">Integrated_hallmarks_mlog10</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="o">-</span><span class="mi">5</span><span class="p">:]</span> <span class="c1"># sort by -log10(P value)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">top_pathways</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;barh&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#649dcd&#39;</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">hatch</span><span class="o">=</span><span class="s1">&#39;///&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Top hallmark gene sets&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;-log$_</span><span class="si">{10}</span><span class="s2">$(P adjusted)&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-2_69_0.png" class="no-scaled-link" src="_images/Tutorial-2_69_0.png" style="width: 642px; height: 395px;" />
</div>
</div>
</div>
<div class="section" id="Key-takeaways">
<h2>Key takeaways<a class="headerlink" href="#Key-takeaways" title="Permalink to this heading"></a></h2>
<p>This Tutorial describes how to generate a gene expression signature by exploiting the inter-operatibility between <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> and <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> and showed how to infer protein activity in single cells from multiple populations with the <a class="reference external" href="https://www.nature.com/articles/s41467-018-03843-3">metaVIPER</a> approach, that leverages multiple lineage-specific regulatory networks. We conclude this Tutorial by presenting another key module used to perform pathway enrichment analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Tutorial-1.html" class="btn btn-neutral float-left" title="Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Tutorial-3.html" class="btn btn-neutral float-right" title="Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, ..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>