<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level &mdash; pyviper 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach" href="Tutorial-2.html" />
    <link rel="prev" title="pyviper package" href="pyviper.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Install-PyVIPER">Install PyVIPER</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Import-modules">Import modules</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Method-1---protein-activity-inference-using-aREA">Method 1 - protein activity inference using aREA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Method-2---protein-activity-inference-using-(matrix)-NaRnEA">Method 2 - protein activity inference using (matrix)-NaRnEA</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-3.html">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Tutorial-1.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="Tutorial-1---Analyzing-scRNA-seq-data-at-the-Protein-Activity-Level">
<h1>Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level<a class="headerlink" href="#Tutorial-1---Analyzing-scRNA-seq-data-at-the-Protein-Activity-Level" title="Permalink to this heading"></a></h1>
<p>This Tutorial will guide the user thorugh the analysis of single-cell RNA-seq (scRNA-seq) data at the Protein Activity level using <code class="docutils literal notranslate"><span class="pre">pyviper</span></code>. Specifically, we will show how to transform a gene expression signature generated from sparse scRNA-seq data into a protein activity matrix using the <a class="reference external" href="https://www.nature.com/articles/ng.3593">VIPER</a> algorithm. After a brief description of the installation procedure and the modules needed, this Jupyter Notebook is organized in the following sections.</p>
<p><strong>Table of Contents</strong></p>
<div class="line-block">
<div class="line">Step 1. Load a gene expression signature for single cells</div>
<div class="line">Step 2. Load and inspect a lineage-specific gene regulatory network</div>
<div class="line">Step 3. Convert the gene expression signature into a protein activity matrix using VIPER</div>
<div class="line">Step 4. Analyze single-cells at the Protein Activity level</div>
<div class="line">Key Takeaways</div>
</div>
<div class="section" id="Install-PyVIPER">
<h2>Install PyVIPER<a class="headerlink" href="#Install-PyVIPER" title="Permalink to this heading"></a></h2>
<p>Install <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> from PyPI using pip. Alternatively, refer to the README in the current GitHub to install from the local directory.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># !pip install viper-in-python</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Import-modules">
<h2>Import modules<a class="headerlink" href="#Import-modules" title="Permalink to this heading"></a></h2>
<p>Load <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> and additional modules required used in this tutorial.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyviper</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;.*The &#39;nopython&#39; keyword.*&quot;</span><span class="p">)</span> <span class="c1"># for jit decorator issue with sc.pp.neighbors (09/30/2023)</span>
</pre></div>
</div>
</div>
<p>Setup the path to the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_location</span> <span class="o">=</span> <span class="s2">&quot;https://zenodo.org/records/10059791/files/&quot;</span>
</pre></div>
</div>
</div>
<p>We will import the data directly from the above link. Optionally, users are allowed to download the files from <a class="reference external" href="https://zenodo.org/records/10059791">here</a> and to load them from their local computer. If the second option is preferred, the following lines can be uncommented and run to download the data on Unix systems and set the working directory to the path where the data will be stored.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!mkdir Tutorial_1_data_dir</span>
<span class="c1">#%cd Tutorial_1_data_dir</span>
<span class="c1">#!curl -O https://zenodo.org/records/10059791/files/Tutorial_1_gExpr_fibroblast_5802.tsv.gz -O https://zenodo.org/records/10059791/files/fibroblast-net.tsv</span>
<span class="c1">#data_location = !pwd</span>
<span class="c1">#data_location = data_location[0]+&quot;/&quot; # new value for data if working on data downloaded locally</span>
</pre></div>
</div>
</div>
<p>### Step 1. Load a gene expression signature for single-cells Load a gene expression signature to be used as input for the <code class="docutils literal notranslate"><span class="pre">viper</span></code> function for Protein Activity inference and store it as an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">AnnData</a> object to enable interoperability with <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/#">scanpy</a>. The time required for this process is approximately 5 s and 30 s if loading from the local directory or remote url, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_expr_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;Tutorial_1_gExpr_fibroblast_5802.tsv.gz&quot;</span>

<span class="n">gene_expr_signature</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_expr_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># read from remote</span>
<span class="n">gene_expr_signature</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">gene_expr_signature</span><span class="p">)</span> <span class="c1"># convert to AnnData object</span>
</pre></div>
</div>
</div>
<p>This signature was generated from a population of cancer-asssociated fibroblasts (CAFs) from publicly avaiable data from <a class="reference external" href="https://www.nature.com/articles/s41422-019-0195-y">Peng et al., 2019</a> following a similar procedure to the one described in Tutorial 2.</p>
<p>Display matrix dimensions with cells on rows and features (genes) on columns.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_expr_signature</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 5802 × 5000
</pre></div></div>
</div>
<p>### Step 2. Load an inspect a lineage-specific gene regulatory network Load and inspect a lineage-specific gene regulatory network generated with the <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">ARACNe3</a>. The time required to load a network is approximately 1 s (from local directory) and 40 s (from remote url).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_path</span> <span class="o">=</span> <span class="n">data_location</span> <span class="o">+</span> <span class="s2">&quot;fibroblast-net.tsv&quot;</span> <span class="c1"># path to ARACNe network</span>

<span class="n">network</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">network_path</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Convert the pd.DataFrame to an object of class <code class="docutils literal notranslate"><span class="pre">Interactome</span></code> to enable easier manipulation. Show the number of regulons using the <code class="docutils literal notranslate"><span class="pre">size()</span></code> method. You can see all available methods by typing <code class="docutils literal notranslate"><span class="pre">help(pyviper.Interactome)</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_interactome</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">Interactome</span><span class="p">(</span><span class="s1">&#39;fibroblasts&#39;</span><span class="p">,</span> <span class="n">network</span><span class="p">)</span> <span class="c1"># convert to class Interactome</span>
<span class="n">network_interactome</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="c1"># show number of regulons in the Interactome</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
6437
</pre></div></div>
</div>
<p>Display the first 5 interactions in the regulatory network. <code class="docutils literal notranslate"><span class="pre">mor</span></code> and <code class="docutils literal notranslate"><span class="pre">likelihood</span></code> are the mode of regulation of the given regulator-target pair and the likelihood of the interaction, respectively. Algorithms other than ARACNe can be used to generate gene regulatory networks, but we recommend postprocessing the output to suitable data formats to use <code class="docutils literal notranslate"><span class="pre">viper</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_interactome</span><span class="o">.</span><span class="n">net_table</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>regulator</th>
      <th>target</th>
      <th>mor</th>
      <th>likelihood</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AATF</td>
      <td>CCM2</td>
      <td>0.012159</td>
      <td>0.999799</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AATF</td>
      <td>THOC2</td>
      <td>0.007487</td>
      <td>0.999597</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AATF</td>
      <td>RAP1A</td>
      <td>-0.021062</td>
      <td>0.999396</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AATF</td>
      <td>SFT2D1</td>
      <td>0.019400</td>
      <td>0.999195</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AATF</td>
      <td>DNAJC2</td>
      <td>0.047129</td>
      <td>0.998993</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Filter out targets in the <code class="docutils literal notranslate"><span class="pre">Interactome</span></code> that are not present in the gene expression matrix. This step is recommended before running <code class="docutils literal notranslate"><span class="pre">viper</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_interactome</span><span class="o">.</span><span class="n">filter_targets</span><span class="p">(</span><span class="n">gene_expr_signature</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Removed 3047484 targets.
</pre></div></div>
</div>
<p>As an example, display the number of targets of two regulators, MYC and EGFR2.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_MYC</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">network_interactome</span><span class="o">.</span><span class="n">get_reg</span><span class="p">(</span><span class="s1">&#39;MYC&#39;</span><span class="p">))</span> <span class="c1"># number of MYC targets in the network</span>
<span class="n">n_EGFR</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">network_interactome</span><span class="o">.</span><span class="n">get_reg</span><span class="p">(</span><span class="s1">&#39;EGFR&#39;</span><span class="p">))</span> <span class="c1"># number of SERPINA12 targets in the network</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The number of targets of MYC and SERPINA12 are &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_MYC</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; and &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_EGFR</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, respectively.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The number of targets of MYC and SERPINA12 are 284 and 183, respectively.
</pre></div></div>
</div>
<p>### Step 3. Convert the gene expression signature into a protein activity matrix using VIPER <code class="docutils literal notranslate"><span class="pre">viper</span></code> transforms a gene expression signature into a protein activity matrix by performing enrichment analysis on regulons. Two methods for enrichment analysis are available, <a class="reference external" href="https://www.nature.com/articles/ng.3593">aREA</a> (default) and <a class="reference external" href="https://www.mdpi.com/1099-4300/25/3/542">NaRnEA</a> (by setting <code class="docutils literal notranslate"><span class="pre">enrichment=&quot;narnea&quot;</span></code>).</p>
<div class="section" id="Method-1---protein-activity-inference-using-aREA">
<h3>Method 1 - protein activity inference using aREA<a class="headerlink" href="#Method-1---protein-activity-inference-using-aREA" title="Permalink to this heading"></a></h3>
<p>As shown above, different regulators can have a different number of targets. Prune each regulon to have the same number of targets (50 in this case). This preprocessing step is advisable to avoid regulators with exceedingly number of targets dominating those having fewer in the computation of Normalized Enrichment Scores (NES), to make NES comparable.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_pruned</span> <span class="o">=</span> <span class="n">network_interactome</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># generate a copy of the unpruned network before pruning it</span>
<span class="n">network_pruned</span><span class="o">.</span><span class="n">prune</span><span class="p">(</span><span class="n">max_targets</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">eliminate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># prune interactome to have exactly 50 targets</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Removed 664967 targets.
Removed 47 regulators.
</pre></div></div>
</div>
<p>Now all the regulators in the network have exactly 50 transcriptional targets. See e.g. MYC and EGFR2.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_MYC</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">network_pruned</span><span class="o">.</span><span class="n">get_reg</span><span class="p">(</span><span class="s1">&#39;MYC&#39;</span><span class="p">))</span> <span class="c1"># number of MYC targets in the network</span>
<span class="n">EGFR</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">network_pruned</span><span class="o">.</span><span class="n">get_reg</span><span class="p">(</span><span class="s1">&#39;EGFR&#39;</span><span class="p">))</span> <span class="c1"># number of SERPINA12 targets in the network</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of MYC targets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_MYC</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Number of EGFR targets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">EGFR</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of MYC targets: 50
Number of EGFR targets: 50
</pre></div></div>
</div>
<div class="alert alert-block alert-success"><p>Run <code class="docutils literal notranslate"><span class="pre">viper</span></code> to compute the protein activity matrix with aREA.</p>
</div><p>Mandatory inputs to <code class="docutils literal notranslate"><span class="pre">viper</span></code> are 1. a gene expression signature and 2. a gene regulatory network. Set the <code class="docutils literal notranslate"><span class="pre">viper</span></code> output to be an ndarray (The default output would be an AnnData object). The expected runtime is approximately 10-15 s.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_aREA</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">gex_data</span><span class="o">=</span><span class="n">gene_expr_signature</span><span class="p">,</span> <span class="c1"># gene expression signature</span>
                             <span class="n">interactome</span><span class="o">=</span><span class="n">network_pruned</span><span class="p">,</span> <span class="c1"># gene regulatory network</span>
                             <span class="n">enrichment</span> <span class="o">=</span> <span class="s2">&quot;area&quot;</span><span class="p">,</span>
                             <span class="n">output_as_anndata</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                             <span class="n">njobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                             <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">ProtAct_aREA</span></code> contains the activity of every regulatory protein described by its NES for each single cell.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_aREA</span> <span class="c1"># display the protein activity matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>regulator</th>
      <th>AAMP</th>
      <th>AATF</th>
      <th>ABCA1</th>
      <th>ABCA12</th>
      <th>ABCA3</th>
      <th>ABCA7</th>
      <th>ABCA8</th>
      <th>ABCB1</th>
      <th>ABCB4</th>
      <th>ABCB5</th>
      <th>...</th>
      <th>ZSWIM5</th>
      <th>ZSWIM6</th>
      <th>ZSWIM7</th>
      <th>ZWINT</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZYX</th>
      <th>ZZEF1</th>
      <th>ZZZ3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AAACCTGAGATGTCGG</th>
      <td>-0.014206</td>
      <td>0.027712</td>
      <td>0.057313</td>
      <td>0.095473</td>
      <td>-0.021236</td>
      <td>0.038511</td>
      <td>-0.598815</td>
      <td>-1.505867</td>
      <td>0.044179</td>
      <td>-0.032513</td>
      <td>...</td>
      <td>0.157645</td>
      <td>0.190605</td>
      <td>-1.006026</td>
      <td>-0.016635</td>
      <td>0.053824</td>
      <td>-0.483114</td>
      <td>-0.164187</td>
      <td>0.598380</td>
      <td>-0.219886</td>
      <td>0.042218</td>
    </tr>
    <tr>
      <th>T1_AACTGGTTCTAACGGT</th>
      <td>-0.251498</td>
      <td>0.049527</td>
      <td>0.058710</td>
      <td>0.086291</td>
      <td>-0.130560</td>
      <td>0.008993</td>
      <td>-0.170331</td>
      <td>-0.038688</td>
      <td>-0.033160</td>
      <td>-0.023936</td>
      <td>...</td>
      <td>0.050540</td>
      <td>-0.001874</td>
      <td>0.852565</td>
      <td>-0.003012</td>
      <td>0.046448</td>
      <td>-0.073733</td>
      <td>0.650326</td>
      <td>1.442805</td>
      <td>0.026046</td>
      <td>-0.014535</td>
    </tr>
    <tr>
      <th>T1_AAGGTTCGTGTCAATC</th>
      <td>-0.436400</td>
      <td>0.058650</td>
      <td>0.071995</td>
      <td>0.085530</td>
      <td>0.007212</td>
      <td>-0.037141</td>
      <td>-0.350010</td>
      <td>-0.271247</td>
      <td>0.026917</td>
      <td>-0.005419</td>
      <td>...</td>
      <td>0.588326</td>
      <td>-0.040662</td>
      <td>0.265022</td>
      <td>-0.185548</td>
      <td>-0.007104</td>
      <td>-0.078222</td>
      <td>0.020503</td>
      <td>-0.784129</td>
      <td>-0.116486</td>
      <td>0.008290</td>
    </tr>
    <tr>
      <th>T1_ACAGCCGTCAAGCCTA</th>
      <td>-1.034668</td>
      <td>0.052937</td>
      <td>0.096666</td>
      <td>0.110275</td>
      <td>-0.019219</td>
      <td>-0.021259</td>
      <td>-0.021779</td>
      <td>-0.703762</td>
      <td>0.007190</td>
      <td>-0.045264</td>
      <td>...</td>
      <td>0.167126</td>
      <td>0.056834</td>
      <td>0.557881</td>
      <td>-0.548752</td>
      <td>-0.010246</td>
      <td>-0.044021</td>
      <td>-0.007267</td>
      <td>0.501714</td>
      <td>-0.001597</td>
      <td>-0.007290</td>
    </tr>
    <tr>
      <th>T1_AGACGTTAGGCGACAT</th>
      <td>-0.811760</td>
      <td>0.043792</td>
      <td>0.066080</td>
      <td>0.069523</td>
      <td>-0.035705</td>
      <td>-0.035428</td>
      <td>-0.317873</td>
      <td>0.037946</td>
      <td>-0.005969</td>
      <td>-0.024018</td>
      <td>...</td>
      <td>0.104896</td>
      <td>-0.012483</td>
      <td>-0.457026</td>
      <td>0.211869</td>
      <td>-0.144309</td>
      <td>-0.019348</td>
      <td>-0.035585</td>
      <td>0.085054</td>
      <td>-0.352943</td>
      <td>-0.011871</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCAGACAAAT</th>
      <td>0.002299</td>
      <td>0.013470</td>
      <td>0.060106</td>
      <td>1.236346</td>
      <td>0.018945</td>
      <td>0.028562</td>
      <td>-0.093820</td>
      <td>-1.275721</td>
      <td>-1.226240</td>
      <td>0.386276</td>
      <td>...</td>
      <td>-0.031268</td>
      <td>-0.062835</td>
      <td>-0.006546</td>
      <td>0.019039</td>
      <td>0.068665</td>
      <td>-0.544908</td>
      <td>0.573362</td>
      <td>-0.551146</td>
      <td>-0.000499</td>
      <td>-0.129102</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCATTAACCG</th>
      <td>-0.129001</td>
      <td>0.034863</td>
      <td>0.089672</td>
      <td>0.044609</td>
      <td>0.040947</td>
      <td>-0.078091</td>
      <td>-0.200762</td>
      <td>-0.007241</td>
      <td>0.042752</td>
      <td>-0.070949</td>
      <td>...</td>
      <td>0.024643</td>
      <td>-0.484168</td>
      <td>-0.022466</td>
      <td>-0.023258</td>
      <td>-0.045327</td>
      <td>-0.041769</td>
      <td>-0.025815</td>
      <td>-0.003803</td>
      <td>-0.054201</td>
      <td>-0.011256</td>
    </tr>
    <tr>
      <th>T24_TTTATGCTCCTTTACA</th>
      <td>-0.354839</td>
      <td>0.036691</td>
      <td>0.062033</td>
      <td>0.054651</td>
      <td>-0.018591</td>
      <td>0.102881</td>
      <td>-0.535426</td>
      <td>-0.809012</td>
      <td>0.000806</td>
      <td>0.072612</td>
      <td>...</td>
      <td>0.074007</td>
      <td>0.091054</td>
      <td>0.080505</td>
      <td>0.003998</td>
      <td>0.030402</td>
      <td>-0.000688</td>
      <td>-0.049335</td>
      <td>-1.073907</td>
      <td>-0.215079</td>
      <td>0.007779</td>
    </tr>
    <tr>
      <th>T24_TTTCCTCCACCGGAAA</th>
      <td>-0.742043</td>
      <td>0.049942</td>
      <td>0.065039</td>
      <td>0.105232</td>
      <td>0.236517</td>
      <td>-0.762427</td>
      <td>-0.290367</td>
      <td>1.586747</td>
      <td>0.020128</td>
      <td>-0.075647</td>
      <td>...</td>
      <td>0.032234</td>
      <td>0.398886</td>
      <td>-0.547757</td>
      <td>-0.016941</td>
      <td>0.260247</td>
      <td>-0.038721</td>
      <td>0.012140</td>
      <td>0.437124</td>
      <td>-0.040129</td>
      <td>-0.040400</td>
    </tr>
    <tr>
      <th>T24_TTTGCGCTCCTTTCGG</th>
      <td>-1.030933</td>
      <td>0.030371</td>
      <td>0.090810</td>
      <td>0.093285</td>
      <td>0.010458</td>
      <td>-0.062967</td>
      <td>0.552246</td>
      <td>-1.137227</td>
      <td>-0.015079</td>
      <td>-0.065197</td>
      <td>...</td>
      <td>-0.782121</td>
      <td>-0.306618</td>
      <td>-0.194269</td>
      <td>0.008829</td>
      <td>-0.011833</td>
      <td>-0.075083</td>
      <td>-0.043519</td>
      <td>0.334202</td>
      <td>-0.869105</td>
      <td>-0.009373</td>
    </tr>
  </tbody>
</table>
<p>5802 rows × 6390 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Method-2---protein-activity-inference-using-(matrix)-NaRnEA">
<h3>Method 2 - protein activity inference using (matrix)-NaRnEA<a class="headerlink" href="#Method-2---protein-activity-inference-using-(matrix)-NaRnEA" title="Permalink to this heading"></a></h3>
<div class="alert alert-block alert-success"><p>Run <code class="docutils literal notranslate"><span class="pre">viper</span></code> to compute the protein activity matrix using (matrix-based) NaRnEA .</p>
</div><p>Compute the protein activity using <code class="docutils literal notranslate"><span class="pre">viper</span></code> with NaRnEA as the enrichment method. Unlike aREA, the NaRnEA framework is suitable also for regulons of different size. Store the output in an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object (default).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">viper</span><span class="p">(</span><span class="n">gex_data</span><span class="o">=</span><span class="n">gene_expr_signature</span><span class="p">,</span> <span class="c1"># gene expression signature</span>
                                 <span class="n">interactome</span><span class="o">=</span><span class="n">network_interactome</span><span class="p">,</span> <span class="c1"># gene regulatory network</span>
                                 <span class="n">enrichment</span> <span class="o">=</span> <span class="s2">&quot;narnea&quot;</span><span class="p">,</span>
                                 <span class="n">njobs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                 <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Show the output type and length of <code class="docutils literal notranslate"><span class="pre">ProtAct_NaRnEA</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span> <span class="c1"># display the protein activity matrix as AnnData object</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 5802 × 6413
    uns: &#39;gex_data&#39;
    layers: &#39;pes&#39;
</pre></div></div>
</div>
<p>When run with the NaRnEA method, <code class="docutils literal notranslate"><span class="pre">viper</span></code> returns an additional layer, <code class="docutils literal notranslate"><span class="pre">pes</span></code> that stores the Proportional Enrichment Score (PES) of each regulator, a measure of protein activity in the <span class="math notranslate nohighlight">\(-1 \le PES \le 1\)</span> interval that can be used as an enhanced measure of effect size of each regulon’s activity. For further details, see <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/36981431/">Griffin et al., 2022</a>.</p>
<p>Display NES protein activity matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span> <span class="c1"># NES matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AAMP</th>
      <th>AATF</th>
      <th>ABCA1</th>
      <th>ABCA12</th>
      <th>ABCA3</th>
      <th>ABCA7</th>
      <th>ABCA8</th>
      <th>ABCB1</th>
      <th>ABCB4</th>
      <th>ABCB5</th>
      <th>...</th>
      <th>ZSWIM5</th>
      <th>ZSWIM6</th>
      <th>ZSWIM7</th>
      <th>ZWINT</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZYX</th>
      <th>ZZEF1</th>
      <th>ZZZ3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AAACCTGAGATGTCGG</th>
      <td>4.715789</td>
      <td>-6.613786</td>
      <td>7.001248</td>
      <td>1.851045</td>
      <td>1.843629</td>
      <td>-0.857881</td>
      <td>14.543342</td>
      <td>-0.613878</td>
      <td>-0.824745</td>
      <td>0.406057</td>
      <td>...</td>
      <td>-1.862346</td>
      <td>3.146378</td>
      <td>-7.006648</td>
      <td>-0.756162</td>
      <td>-2.003027</td>
      <td>1.459852</td>
      <td>2.712799</td>
      <td>5.787629</td>
      <td>2.397046</td>
      <td>-2.582086</td>
    </tr>
    <tr>
      <th>T1_AACTGGTTCTAACGGT</th>
      <td>-7.191714</td>
      <td>7.533881</td>
      <td>9.429433</td>
      <td>4.088336</td>
      <td>-0.461214</td>
      <td>1.055557</td>
      <td>-18.886093</td>
      <td>1.401654</td>
      <td>-0.798170</td>
      <td>1.079553</td>
      <td>...</td>
      <td>2.199012</td>
      <td>-3.213969</td>
      <td>-7.219745</td>
      <td>0.821136</td>
      <td>1.227563</td>
      <td>-1.558459</td>
      <td>-1.394093</td>
      <td>12.747848</td>
      <td>1.822417</td>
      <td>2.953031</td>
    </tr>
    <tr>
      <th>T1_AAGGTTCGTGTCAATC</th>
      <td>8.887137</td>
      <td>-9.449057</td>
      <td>-10.272199</td>
      <td>3.190093</td>
      <td>-1.265178</td>
      <td>0.657827</td>
      <td>20.582304</td>
      <td>0.466253</td>
      <td>-0.916708</td>
      <td>0.486921</td>
      <td>...</td>
      <td>1.655850</td>
      <td>3.406829</td>
      <td>-8.988820</td>
      <td>-1.780032</td>
      <td>-0.947679</td>
      <td>-2.583600</td>
      <td>1.037094</td>
      <td>-11.109167</td>
      <td>-1.955563</td>
      <td>-2.474079</td>
    </tr>
    <tr>
      <th>T1_ACAGCCGTCAAGCCTA</th>
      <td>7.847414</td>
      <td>6.900573</td>
      <td>9.575873</td>
      <td>-2.852121</td>
      <td>-0.872999</td>
      <td>-1.455654</td>
      <td>23.062950</td>
      <td>-0.603498</td>
      <td>1.127638</td>
      <td>0.936162</td>
      <td>...</td>
      <td>-1.165667</td>
      <td>-2.528141</td>
      <td>-6.812299</td>
      <td>-1.115750</td>
      <td>-1.571316</td>
      <td>1.245981</td>
      <td>2.074361</td>
      <td>-11.699137</td>
      <td>2.160041</td>
      <td>2.499193</td>
    </tr>
    <tr>
      <th>T1_AGACGTTAGGCGACAT</th>
      <td>7.643162</td>
      <td>6.953849</td>
      <td>-8.838992</td>
      <td>2.588946</td>
      <td>0.552425</td>
      <td>-0.694912</td>
      <td>-15.233547</td>
      <td>0.163285</td>
      <td>2.939726</td>
      <td>-0.930142</td>
      <td>...</td>
      <td>1.883423</td>
      <td>-2.452158</td>
      <td>-9.295234</td>
      <td>0.852636</td>
      <td>1.995564</td>
      <td>-1.274296</td>
      <td>-2.888875</td>
      <td>8.253686</td>
      <td>-2.171622</td>
      <td>3.023637</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCAGACAAAT</th>
      <td>3.929488</td>
      <td>3.904004</td>
      <td>-7.629382</td>
      <td>-2.257839</td>
      <td>0.547175</td>
      <td>-1.310761</td>
      <td>10.857200</td>
      <td>0.809013</td>
      <td>-1.093720</td>
      <td>1.711279</td>
      <td>...</td>
      <td>-2.431247</td>
      <td>2.670952</td>
      <td>5.755338</td>
      <td>0.977276</td>
      <td>-2.342592</td>
      <td>-0.952300</td>
      <td>1.586434</td>
      <td>6.731536</td>
      <td>2.967126</td>
      <td>1.218217</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCATTAACCG</th>
      <td>-7.197257</td>
      <td>7.793606</td>
      <td>8.194140</td>
      <td>-2.333065</td>
      <td>-1.661859</td>
      <td>1.403345</td>
      <td>-19.088629</td>
      <td>0.623605</td>
      <td>-0.571697</td>
      <td>-1.285721</td>
      <td>...</td>
      <td>1.153792</td>
      <td>-2.933194</td>
      <td>7.263271</td>
      <td>-0.940619</td>
      <td>-1.737071</td>
      <td>1.691363</td>
      <td>-1.825287</td>
      <td>10.659641</td>
      <td>1.579409</td>
      <td>2.971451</td>
    </tr>
    <tr>
      <th>T24_TTTATGCTCCTTTACA</th>
      <td>-7.103387</td>
      <td>8.042507</td>
      <td>-8.372260</td>
      <td>2.609825</td>
      <td>0.896921</td>
      <td>-2.094449</td>
      <td>20.645484</td>
      <td>-0.346492</td>
      <td>0.872927</td>
      <td>-0.579501</td>
      <td>...</td>
      <td>1.109876</td>
      <td>2.796211</td>
      <td>6.435829</td>
      <td>1.046557</td>
      <td>1.433361</td>
      <td>-1.580545</td>
      <td>1.450763</td>
      <td>-10.226545</td>
      <td>1.772312</td>
      <td>2.924802</td>
    </tr>
    <tr>
      <th>T24_TTTCCTCCACCGGAAA</th>
      <td>-6.267298</td>
      <td>-5.975001</td>
      <td>-6.912350</td>
      <td>-3.515971</td>
      <td>0.877991</td>
      <td>-1.020563</td>
      <td>-11.196501</td>
      <td>0.534728</td>
      <td>-1.318767</td>
      <td>0.807623</td>
      <td>...</td>
      <td>1.258743</td>
      <td>-3.328178</td>
      <td>5.359631</td>
      <td>-2.210881</td>
      <td>-1.586171</td>
      <td>-1.064334</td>
      <td>0.944330</td>
      <td>-6.458201</td>
      <td>-1.191704</td>
      <td>2.884055</td>
    </tr>
    <tr>
      <th>T24_TTTGCGCTCCTTTCGG</th>
      <td>7.656701</td>
      <td>5.664298</td>
      <td>-7.455275</td>
      <td>3.304026</td>
      <td>1.240349</td>
      <td>-2.871128</td>
      <td>15.743551</td>
      <td>0.305051</td>
      <td>-1.252826</td>
      <td>1.436090</td>
      <td>...</td>
      <td>-1.513860</td>
      <td>-2.786318</td>
      <td>5.453333</td>
      <td>1.610768</td>
      <td>1.393192</td>
      <td>1.825675</td>
      <td>1.602548</td>
      <td>-6.023640</td>
      <td>1.601451</td>
      <td>2.207284</td>
    </tr>
  </tbody>
</table>
<p>5802 rows × 6413 columns</p>
</div></div>
</div>
<p>Display the PES protein activity matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span><span class="o">.</span><span class="n">to_df</span><span class="p">(</span><span class="n">layer</span><span class="o">=</span><span class="s2">&quot;pes&quot;</span><span class="p">)</span> <span class="c1"># PES matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AAMP</th>
      <th>AATF</th>
      <th>ABCA1</th>
      <th>ABCA12</th>
      <th>ABCA3</th>
      <th>ABCA7</th>
      <th>ABCA8</th>
      <th>ABCB1</th>
      <th>ABCB4</th>
      <th>ABCB5</th>
      <th>...</th>
      <th>ZSWIM5</th>
      <th>ZSWIM6</th>
      <th>ZSWIM7</th>
      <th>ZWINT</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZYX</th>
      <th>ZZEF1</th>
      <th>ZZZ3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AAACCTGAGATGTCGG</th>
      <td>0.240772</td>
      <td>-0.304270</td>
      <td>0.281038</td>
      <td>0.074875</td>
      <td>0.065383</td>
      <td>-0.030209</td>
      <td>0.378750</td>
      <td>-0.021176</td>
      <td>-0.028191</td>
      <td>0.012894</td>
      <td>...</td>
      <td>-0.057990</td>
      <td>0.109255</td>
      <td>-0.352088</td>
      <td>-0.026435</td>
      <td>-0.069684</td>
      <td>0.050031</td>
      <td>0.092210</td>
      <td>0.262122</td>
      <td>0.085055</td>
      <td>-0.092093</td>
    </tr>
    <tr>
      <th>T1_AACTGGTTCTAACGGT</th>
      <td>-0.377572</td>
      <td>0.314127</td>
      <td>0.353572</td>
      <td>0.160783</td>
      <td>-0.015217</td>
      <td>0.035737</td>
      <td>-0.497909</td>
      <td>0.047375</td>
      <td>-0.026517</td>
      <td>0.033253</td>
      <td>...</td>
      <td>0.065166</td>
      <td>-0.115518</td>
      <td>-0.365130</td>
      <td>0.029192</td>
      <td>0.042427</td>
      <td>-0.052575</td>
      <td>-0.044594</td>
      <td>0.546236</td>
      <td>0.062533</td>
      <td>0.098134</td>
    </tr>
    <tr>
      <th>T1_AAGGTTCGTGTCAATC</th>
      <td>0.411746</td>
      <td>-0.426309</td>
      <td>-0.463015</td>
      <td>0.121526</td>
      <td>-0.039884</td>
      <td>0.021507</td>
      <td>0.489378</td>
      <td>0.015299</td>
      <td>-0.029489</td>
      <td>0.014491</td>
      <td>...</td>
      <td>0.047115</td>
      <td>0.107279</td>
      <td>-0.455701</td>
      <td>-0.057125</td>
      <td>-0.030605</td>
      <td>-0.084660</td>
      <td>0.033765</td>
      <td>-0.547947</td>
      <td>-0.066875</td>
      <td>-0.085122</td>
    </tr>
    <tr>
      <th>T1_ACAGCCGTCAAGCCTA</th>
      <td>0.373952</td>
      <td>0.280809</td>
      <td>0.347536</td>
      <td>-0.110712</td>
      <td>-0.028223</td>
      <td>-0.049295</td>
      <td>0.563023</td>
      <td>-0.019842</td>
      <td>0.036926</td>
      <td>0.028396</td>
      <td>...</td>
      <td>-0.035112</td>
      <td>-0.090271</td>
      <td>-0.344878</td>
      <td>-0.036706</td>
      <td>-0.051855</td>
      <td>0.040700</td>
      <td>0.068411</td>
      <td>-0.576989</td>
      <td>0.072889</td>
      <td>0.081400</td>
    </tr>
    <tr>
      <th>T1_AGACGTTAGGCGACAT</th>
      <td>0.367726</td>
      <td>0.285936</td>
      <td>-0.395159</td>
      <td>0.100920</td>
      <td>0.019161</td>
      <td>-0.023652</td>
      <td>-0.399057</td>
      <td>0.005474</td>
      <td>0.096805</td>
      <td>-0.028409</td>
      <td>...</td>
      <td>0.055218</td>
      <td>-0.087718</td>
      <td>-0.469730</td>
      <td>0.030140</td>
      <td>0.068457</td>
      <td>-0.042643</td>
      <td>-0.091376</td>
      <td>0.348384</td>
      <td>-0.075733</td>
      <td>0.099308</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCAGACAAAT</th>
      <td>0.204446</td>
      <td>0.174607</td>
      <td>-0.337181</td>
      <td>-0.093020</td>
      <td>0.019611</td>
      <td>-0.046764</td>
      <td>0.288005</td>
      <td>0.028439</td>
      <td>-0.037907</td>
      <td>0.055128</td>
      <td>...</td>
      <td>-0.076656</td>
      <td>0.094553</td>
      <td>0.273277</td>
      <td>0.035624</td>
      <td>-0.082810</td>
      <td>-0.033380</td>
      <td>0.054557</td>
      <td>0.311536</td>
      <td>0.106925</td>
      <td>0.042919</td>
    </tr>
    <tr>
      <th>T24_TTGCGTCCATTAACCG</th>
      <td>-0.379965</td>
      <td>0.335563</td>
      <td>0.321241</td>
      <td>-0.093680</td>
      <td>-0.056283</td>
      <td>0.048430</td>
      <td>-0.508088</td>
      <td>0.021407</td>
      <td>-0.019326</td>
      <td>-0.040297</td>
      <td>...</td>
      <td>0.034984</td>
      <td>-0.105993</td>
      <td>0.329845</td>
      <td>-0.032393</td>
      <td>-0.059631</td>
      <td>0.057258</td>
      <td>-0.059820</td>
      <td>0.473351</td>
      <td>0.055332</td>
      <td>0.101333</td>
    </tr>
    <tr>
      <th>T24_TTTATGCTCCTTTACA</th>
      <td>-0.374228</td>
      <td>0.343792</td>
      <td>-0.370403</td>
      <td>0.103941</td>
      <td>0.031494</td>
      <td>-0.072713</td>
      <td>0.525797</td>
      <td>-0.011750</td>
      <td>0.029375</td>
      <td>-0.018084</td>
      <td>...</td>
      <td>0.033471</td>
      <td>0.094812</td>
      <td>0.289742</td>
      <td>0.037420</td>
      <td>0.050027</td>
      <td>-0.053926</td>
      <td>0.048746</td>
      <td>-0.502225</td>
      <td>0.061779</td>
      <td>0.099145</td>
    </tr>
    <tr>
      <th>T24_TTTCCTCCACCGGAAA</th>
      <td>-0.332941</td>
      <td>-0.275806</td>
      <td>-0.310930</td>
      <td>-0.140664</td>
      <td>0.031023</td>
      <td>-0.035522</td>
      <td>-0.299154</td>
      <td>0.018338</td>
      <td>-0.044461</td>
      <td>0.025257</td>
      <td>...</td>
      <td>0.037913</td>
      <td>-0.120995</td>
      <td>0.238777</td>
      <td>-0.075475</td>
      <td>-0.054143</td>
      <td>-0.036417</td>
      <td>0.031897</td>
      <td>-0.320772</td>
      <td>-0.042458</td>
      <td>0.097503</td>
    </tr>
    <tr>
      <th>T24_TTTGCGCTCCTTTCGG</th>
      <td>0.387339</td>
      <td>0.245869</td>
      <td>-0.329289</td>
      <td>0.132855</td>
      <td>0.043827</td>
      <td>-0.100557</td>
      <td>0.406450</td>
      <td>0.010523</td>
      <td>-0.042570</td>
      <td>0.045317</td>
      <td>...</td>
      <td>-0.046915</td>
      <td>-0.100964</td>
      <td>0.250039</td>
      <td>0.057934</td>
      <td>0.049015</td>
      <td>0.062157</td>
      <td>0.054237</td>
      <td>-0.295971</td>
      <td>0.056435</td>
      <td>0.075792</td>
    </tr>
  </tbody>
</table>
<p>5802 rows × 6413 columns</p>
</div></div>
</div>
<p>Notice that when setting the <code class="docutils literal notranslate"><span class="pre">viper</span></code> output to be an object of class <code class="docutils literal notranslate"><span class="pre">AnnData</span></code>, the input gene expression matrix is saved under the attribute <code class="docutils literal notranslate"><span class="pre">gex_data</span></code> of the output <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;gex_data&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>RP11-54O7.3</th>
      <th>SAMD11</th>
      <th>PLEKHN1</th>
      <th>RP11-54O7.17</th>
      <th>HES4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AAACCTGAGATGTCGG</th>
      <td>-0.145409</td>
      <td>-0.513375</td>
      <td>-0.130302</td>
      <td>-0.134275</td>
      <td>1.855504</td>
    </tr>
    <tr>
      <th>T1_AACTGGTTCTAACGGT</th>
      <td>-0.108846</td>
      <td>-0.390002</td>
      <td>-0.096722</td>
      <td>-0.101255</td>
      <td>2.064095</td>
    </tr>
    <tr>
      <th>T1_AAGGTTCGTGTCAATC</th>
      <td>-0.084144</td>
      <td>-0.306366</td>
      <td>-0.074026</td>
      <td>-0.078937</td>
      <td>1.387576</td>
    </tr>
    <tr>
      <th>T1_ACAGCCGTCAAGCCTA</th>
      <td>-0.099796</td>
      <td>-0.359357</td>
      <td>-0.088408</td>
      <td>-0.093080</td>
      <td>-0.622520</td>
    </tr>
    <tr>
      <th>T1_AGACGTTAGGCGACAT</th>
      <td>-0.099830</td>
      <td>-0.359472</td>
      <td>-0.088440</td>
      <td>-0.093111</td>
      <td>3.986061</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>### Step 4. Analyze single-cells at the Protein Activity level This section presents some useful postprocessing functions that can be used on the protein activity matrix.</p>
<p>First, set the set of regulatory proteins to be those for the human species. This is step is redundant, since <code class="docutils literal notranslate"><span class="pre">pyviper</span></code> sets <code class="docutils literal notranslate"><span class="pre">&quot;human&quot;</span></code> as default, but it demonstrates how to have complete control over the analysis, if fine-tuning is desired.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pyviper</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">set_regulators_species_to_use</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Run PCA to reduce the dimensionality of the dataset, by projecting the PES of transcriptional regulators (TFs) and co-transcriptional regulators (coTFs) by setting <code class="docutils literal notranslate"><span class="pre">filter_by_feature_groups=[&quot;tfs&quot;,</span> <span class="pre">&quot;cotfs&quot;]</span></code>. In this example, we focus on TFs and coTFs, major mechanistic determinants of cell state as shown in <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0092867420316172?via%3Dihub">Paull et al., 2021</a> and <a class="reference external" href="https://www.nature.com/articles/nrc.2016.124">Califano &amp; Alvarez, 2016</a>.
Possible options for this parameter are <code class="docutils literal notranslate"><span class="pre">[&quot;tfs&quot;,</span> <span class="pre">&quot;cotfs&quot;,</span> <span class="pre">&quot;sig&quot;,</span> <span class="pre">&quot;surf&quot;]</span></code>, where the latter refer to signaling proteins and surface markers, respectively. Set <code class="docutils literal notranslate"><span class="pre">None</span></code> to use all the features in the protein activity matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">pyviper</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;pes&quot;</span><span class="p">,</span> <span class="n">filter_by_feature_groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tfs&quot;</span><span class="p">,</span><span class="s2">&quot;cotfs&quot;</span><span class="p">],</span> <span class="n">zero_center</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(5802, 2164)
</pre></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">tl</span></code> module provides wrappers to several <code class="docutils literal notranslate"><span class="pre">scanpy.tl</span></code> functions, and it can be used to simplify the postprocessing of protein activity data, for instance by subsetting the class of regulatory proteins to analyze, as shown in the cell above.</p>
<p>Compute the neighbors graph of cells using the PCA representation of the protein activity matrix. For sake of simplicity, set <code class="docutils literal notranslate"><span class="pre">n_neighbors</span> <span class="pre">=</span> <span class="pre">20</span></code>and <code class="docutils literal notranslate"><span class="pre">n_pcs</span> <span class="pre">=</span> <span class="pre">50</span></code> (please, notice that typically, these parameters need tuning).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;correlation&quot;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Cluster cells based on their protein activity profiles using the Leiden algorithm with <code class="docutils literal notranslate"><span class="pre">resolution=0.1</span></code> (in real-world applications this parameter should also be optimized).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">n_iterations</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Generate a <a class="reference external" href="https://arxiv.org/abs/1802.03426">UMAP</a> embedding for the graph.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Display the 2-dimensional UMAP embedding using the <code class="docutils literal notranslate"><span class="pre">pyviper.pl</span></code> module for easier post-processing of key protein activity features. Color by cluster.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="n">n_cells</span> <span class="o">=</span> <span class="n">ProtAct_NaRnEA</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">dot_size</span> <span class="o">=</span> <span class="mi">120000</span> <span class="o">/</span> <span class="n">n_cells</span><span class="o">*</span><span class="mi">2</span>
<span class="n">pyviper</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set1&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">dot_size</span><span class="p">,</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-1_54_0.png" class="no-scaled-link" src="_images/Tutorial-1_54_0.png" style="width: 364px; height: 333px;" />
</div>
</div>
<p>Identify the most active regulators in each cluster by cluster-based integration of their NES using the Stouffer’s method:</p>
<div class="math notranslate nohighlight">
\[NES^C_j = \frac{\sum_{k \in C}NES_{k,j}}{\sqrt{n^C}}\]</div>
<p>where <span class="math notranslate nohighlight">\(NES^C_j\)</span> denotes the integrated NES for regulator <span class="math notranslate nohighlight">\(j\)</span> in cluster <span class="math notranslate nohighlight">\(C\)</span>, <span class="math notranslate nohighlight">\(NES_{k,j}\)</span> is the NES of regulator <span class="math notranslate nohighlight">\(j\)</span> in cluster <span class="math notranslate nohighlight">\(C\)</span> and <span class="math notranslate nohighlight">\(n^C\)</span> is the number of cells in cluster <span class="math notranslate nohighlight">\(C\)</span>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ProtAct_NaRnEA</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">stouffer</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">filter_by_feature_groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tfs&quot;</span><span class="p">,</span><span class="s2">&quot;cotfs&quot;</span><span class="p">])</span> <span class="c1"># Stouffer integration</span>
</pre></div>
</div>
</div>
<p>Display the integrated NES scores in each cluster, are stored in <code class="docutils literal notranslate"><span class="pre">.uns</span></code> under<code class="docutils literal notranslate"><span class="pre">stouffer</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">NES_integrated</span> <span class="o">=</span> <span class="n">ProtAct_NaRnEA</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;stouffer&quot;</span><span class="p">]</span>
<span class="n">NES_integrated</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AATF</th>
      <th>ABCG1</th>
      <th>ABCG4</th>
      <th>ABLIM3</th>
      <th>ABT1</th>
      <th>ACAD8</th>
      <th>ACTN2</th>
      <th>ACTR5</th>
      <th>ADH1A</th>
      <th>ADNP</th>
      <th>...</th>
      <th>ZSWIM1</th>
      <th>ZSWIM3</th>
      <th>ZSWIM4</th>
      <th>ZSWIM5</th>
      <th>ZSWIM6</th>
      <th>ZSWIM7</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZZEF1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>66.091798</td>
      <td>115.405773</td>
      <td>26.758490</td>
      <td>-28.100281</td>
      <td>-132.191396</td>
      <td>-33.599384</td>
      <td>-13.772651</td>
      <td>-47.460612</td>
      <td>4.015370</td>
      <td>-30.219543</td>
      <td>...</td>
      <td>-11.137375</td>
      <td>-22.225285</td>
      <td>61.797696</td>
      <td>20.534119</td>
      <td>-66.629786</td>
      <td>-29.674837</td>
      <td>-11.166953</td>
      <td>1.518258</td>
      <td>-26.212789</td>
      <td>-22.827012</td>
    </tr>
    <tr>
      <th>1</th>
      <td>-95.484010</td>
      <td>-144.304612</td>
      <td>-16.952406</td>
      <td>-17.652945</td>
      <td>19.505636</td>
      <td>36.045797</td>
      <td>-0.281349</td>
      <td>-34.882520</td>
      <td>11.360462</td>
      <td>-158.717215</td>
      <td>...</td>
      <td>29.614091</td>
      <td>26.530288</td>
      <td>-36.345481</td>
      <td>10.419244</td>
      <td>-52.724309</td>
      <td>70.003267</td>
      <td>22.476657</td>
      <td>-9.414996</td>
      <td>21.718254</td>
      <td>26.721121</td>
    </tr>
    <tr>
      <th>2</th>
      <td>160.694616</td>
      <td>112.503595</td>
      <td>-28.526969</td>
      <td>-12.598789</td>
      <td>273.728607</td>
      <td>35.077632</td>
      <td>-18.162602</td>
      <td>0.486704</td>
      <td>11.013269</td>
      <td>-159.268514</td>
      <td>...</td>
      <td>20.997878</td>
      <td>-30.428877</td>
      <td>-3.913593</td>
      <td>-4.204891</td>
      <td>-78.140381</td>
      <td>52.720958</td>
      <td>3.642006</td>
      <td>-28.206880</td>
      <td>23.372432</td>
      <td>-44.491621</td>
    </tr>
    <tr>
      <th>3</th>
      <td>59.755942</td>
      <td>-6.759463</td>
      <td>-18.311627</td>
      <td>-13.108654</td>
      <td>45.268690</td>
      <td>13.422040</td>
      <td>-6.960340</td>
      <td>-5.065040</td>
      <td>4.802130</td>
      <td>-35.830086</td>
      <td>...</td>
      <td>14.509917</td>
      <td>-0.621032</td>
      <td>6.271539</td>
      <td>5.765715</td>
      <td>-7.064015</td>
      <td>41.651426</td>
      <td>15.331487</td>
      <td>-15.428624</td>
      <td>22.443587</td>
      <td>-13.317803</td>
    </tr>
  </tbody>
</table>
<p>4 rows × 2164 columns</p>
</div></div>
</div>
<p>Analogously, the function <code class="docutils literal notranslate"><span class="pre">stouffer_clusters_df</span></code> can be used to performed the same operations on pd.DataFrames. Remove ribosomal proteins from the Stouffer-integrated matrix (ribosomal protein genes are among the most highly expressed genes in most cell types).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">filtered_proteins</span> <span class="o">=</span> <span class="n">NES_integrated</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">~</span><span class="n">NES_integrated</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;RP&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span> <span class="c1"># Exclude columns starting with &#39;RP&#39; from integrated matrix</span>
<span class="n">NES_integrated</span> <span class="o">=</span> <span class="n">NES_integrated</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">filtered_proteins</span><span class="p">]</span> <span class="c1"># remove ribosomal proteins from DataFrame</span>
</pre></div>
</div>
</div>
<p>Select the 10 most activated regulatory protein in each cluster.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">active_proteins</span> <span class="o">=</span> <span class="n">NES_integrated</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># the top 10 most activated proteins in each cell type</span>
<span class="n">active_proteins</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0    [INHBA, MMP14, ID3, LOXL2, HOPX, RUNX2, LEF1, ...
1    [TSHZ2, NFIA, NFIB, ZBTB16, TCF21, NR2F1, ZFP3...
2    [ENO1, PFDN5, NACA, LOXL2, BTF3, HMGA1, CCDC85...
3    [FOSB, NR4A1, ELL2, ID4, REL, JUND, NR4A2, NR4...
dtype: object
</pre></div></div>
</div>
<p>Finally, display the top activated proteins from <code class="docutils literal notranslate"><span class="pre">ProtAct_NaRnEA</span></code> on a heatmap</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protein_set</span> <span class="o">=</span> <span class="n">active_proteins</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="c1"># protein set to display on the heatmap</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">ProtAct_NaRnEA</span><span class="p">,</span> <span class="n">var_names</span><span class="o">=</span><span class="n">protein_set</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">vcenter</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;RdBu_r&quot;</span><span class="p">,</span><span class="n">dendrogram</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">swap_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_gene_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-1_64_0.png" class="no-scaled-link" src="_images/Tutorial-1_64_0.png" style="width: 832px; height: 556px;" />
</div>
</div>
<p>### Key takeaways This Tutorial describes how to inspect and process gene regulatory networks, by manipulating objects of class <code class="docutils literal notranslate"><span class="pre">Interactome</span></code>, and shows how to generate protein activity profiles from single-cell data using the <code class="docutils literal notranslate"><span class="pre">viper</span></code> function. Last, we presented key functions for the post-processing and visualization of protein activity data. In Tutorial 2 we will introduce the user to the generation of suitable gene expression signatures for <code class="docutils literal notranslate"><span class="pre">viper</span></code> and extend the VIPER analysis to
simultaneously account for multiple cell pupulations through the exploitation of the <a class="reference external" href="https://www.nature.com/articles/s41467-018-03843-3">metaVIPER</a> paradigm.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pyviper.html" class="btn btn-neutral float-left" title="pyviper package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Tutorial-2.html" class="btn btn-neutral float-right" title="Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, ..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>