<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis &mdash; pyviper 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach" href="Tutorial-2.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyviper
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyviper.html">pyviper package</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Tutorial-1.html">Tutorial 1 - Analyzing scRNA-seq data at the Protein Activity Level</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial-2.html">Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Install-Pyviper">Install Pyviper</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Import-modules">Import modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">Step 1. Load a gene expression matrix and associated metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-2.-Perform-basic-preprocessing-and-clustering">Step 2. Perform basic preprocessing and clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#3.-Generate-Metacells">3. Generate Metacells</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyviper</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Tutorial-3.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="Tutorial-3---Generating-Metacells-for-ARACNe3-network-generation-and-VIPER-protein-activity-analysis">
<h1>Tutorial 3 - Generating Metacells for ARACNe3 network generation and VIPER protein activity analysis<a class="headerlink" href="#Tutorial-3---Generating-Metacells-for-ARACNe3-network-generation-and-VIPER-protein-activity-analysis" title="Permalink to this heading"></a></h1>
<p>This tutorial explores different approaches of generating metacells for a scRNA-seq dataset. Metacells can potentially reduce sparseness in single-cell transcriptomics and improve protein activity inference and computational efficiency. For ARACNe3’s mutual information estimation, it is vital to have metacells that are independent and identically distributed. Therefore, we have implemented a weighted random sampling approach with a probability decay to disincentivize re-using the same
single-cell expression profiles to create multiple metacells.</p>
<div class="section" id="Install-Pyviper">
<h2>Install Pyviper<a class="headerlink" href="#Install-Pyviper" title="Permalink to this heading"></a></h2>
<p>Install pyviper from PyPI using pip. Alternatively, refer to the README in the current GitHub to install from the local directory.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip install viper-in-python</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Import-modules">
<h2>Import modules<a class="headerlink" href="#Import-modules" title="Permalink to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyviper</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">silhouette_samples</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;.*The &#39;nopython&#39; keyword.*&quot;</span><span class="p">)</span> <span class="c1"># for jit decorator issue with sc.pp.neighbors (09/30/2023)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-1.-Load-a-gene-expression-matrix-and-associated-metadata">
<h2>Step 1. Load a gene expression matrix and associated metadata<a class="headerlink" href="#Step-1.-Load-a-gene-expression-matrix-and-associated-metadata" title="Permalink to this heading"></a></h2>
<p>Load the same gene expression matrix (UMIs) as tutorial 2 and store it into an AnnData object to enable interoperability with scanpy. Cells used in this tutorial were sampled from scRNA-seq data published in Peng et al., 2019.</p>
<p>Display matrix dimensions (cells x genes)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gExpr_url</span> <span class="o">=</span> <span class="s2">&quot;https://zenodo.org/records/10059791/files/Tutorial_2_counts_mixed_4632.tsv.gz&quot;</span> <span class="c1"># path to gene expression matrix (UMI counts)</span>
<span class="n">adata_gExpr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gExpr_url</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># read from remote</span>
<span class="n">adata_gExpr</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span> <span class="c1"># convert to AnnData object</span>
</pre></div>
</div>
</div>
<p>Load cell-associated metadata. Metadata are stored on GitHub.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata_url</span> <span class="o">=</span> <span class="s2">&quot;https://zenodo.org/records/10059791/files/Tutorial_2_metadata_mixed_4632.tsv.gz&quot;</span> <span class="c1"># path to cells metadata</span>
<span class="n">cells_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata_url</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># load it</span>
</pre></div>
</div>
</div>
<p>Store the metadata in the adata_gExpr object as observation annotation.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">cells_metadata</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span><span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># store cell-specific metadata as annotation observation</span>
</pre></div>
</div>
</div>
<p>Display the observation annotation from the AnnData object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell_Type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>T1_AACCATGCACAACTGT</th>
      <td>Ductal cell type 2</td>
    </tr>
    <tr>
      <th>T1_AAGACCTAGTCATGCT</th>
      <td>Ductal cell type 2</td>
    </tr>
    <tr>
      <th>T1_ACATACGAGACTCGGA</th>
      <td>Ductal cell type 2</td>
    </tr>
    <tr>
      <th>T1_ACCAGTATCTTGCAAG</th>
      <td>Ductal cell type 2</td>
    </tr>
    <tr>
      <th>T1_ACGATGTTCACGAAGG</th>
      <td>Ductal cell type 2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The observation annotations include the patient from which each cell was sequenced and the annotated cell type. List the cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Cell_Type&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;n&#39;</span><span class="p">)</span> <span class="c1"># show cell types and number of cells for each type in AnnData</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell_Type</th>
      <th>n</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>B cell</td>
      <td>1000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Ductal cell type 2</td>
      <td>1455</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Fibroblast cell</td>
      <td>1277</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Stellate cell</td>
      <td>900</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Step-2.-Perform-basic-preprocessing-and-clustering">
<h2>Step 2. Perform basic preprocessing and clustering<a class="headerlink" href="#Step-2.-Perform-basic-preprocessing-and-clustering" title="Permalink to this heading"></a></h2>
<p>First, we can perform some preprocessing. The UMI matrix can be processed using the standard Scanpy preprocessing workflow. Since the quality of the provided data was pre-assessed and found to be high, cell filtering will be minimal. For a more detailed explanation of QC steps, refer to the preprocessing tutorials at Scanpy or Seurat.</p>
<p>If we want to generate metacells for each cell type, we eventually need to generate a nearest neighbor graph and then compute silouhette scores.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span> <span class="c1"># filter out cells with &lt;200 genes expressed</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># filter out genes that are detected in &lt;3 cells</span>

<span class="c1">#adata_gExpr = calcMitoRiboPercent(adata_gExpr)</span>
<span class="c1">#adata_gExpr = filterCellsHighMito(adata_gExpr,0.2)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat&quot;</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">adata_gExpr</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata_gExpr</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span>
<br/><br/></pre></div>
</div>
</div>
</div>
<div class="section" id="3.-Generate-Metacells">
<h2>3. Generate Metacells<a class="headerlink" href="#3.-Generate-Metacells" title="Permalink to this heading"></a></h2>
<p>First, we will compute silhouette scores for the cells using their principal components. Then, we will pass the Anndata object with silhouette scores into pyviper’s function generating metacells. This function uses a probability re-weighting approach to incentivize different minimal overlap of single cells in multiple metacells. This probability decay approach can be used with the “use_decay” and “decay_factor” parameters.</p>
<p>The new dataset with generated metacells will be stored in experiment_dir_path under the subfolder with_penalty or without_penalty, depending on whether use_decay is set to True or False, respectively. For each folder, there will be further subfolders for n_cells_per_metacell. This is to aid organization, especially since one could want to compare results across different number of cells per metacell or the use of penalty.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">666</span><span class="p">)</span>
<span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;silhouette_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">silhouette_samples</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">],</span> <span class="n">labels</span><span class="o">=</span><span class="n">adata_gExpr</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">])</span>
<br/><br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_metacell</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">generateMetacellAnnData</span><span class="p">(</span><span class="n">arg0</span><span class="o">=</span><span class="n">adata_gExpr</span><span class="p">,</span>
                                         <span class="n">method</span><span class="o">=</span><span class="s2">&quot;relative&quot;</span><span class="p">,</span>
                                         <span class="n">cluster_label</span><span class="o">=</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span>
                                         <span class="n">n_metacells</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="c1">#not used</span>
                                         <span class="n">n_metacells_per_cluster</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
                                         <span class="n">n_neighbors</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span>
                                         <span class="n">n_cells_per_metacell</span><span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
                                         <span class="n">experiment_dir_path</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                                         <span class="n">use_decay</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                         <span class="n">decay_factor</span><span class="o">=</span> <span class="mf">0.10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Good quality cells  1000
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "62da51ecc51a4aae8090451c9dc7c437", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
Good quality cells  1455
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "769a257bb8944883b11a650cabe6fef9", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
Good quality cells  1277
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d989511299db47fcb34601b2dc67a496", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
(255, 20197)
Good quality cells  900
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "523011a962e24763b915c699203c1fff", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
(255, 20197)
(180, 20197)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/av2729/opt/anaconda3/envs/single-cell-env-scvi/lib/python3.8/site-packages/scanpy/preprocessing/_normalization.py:197: UserWarning: Some cells have zero counts
  warn(UserWarning(&#39;Some cells have zero counts&#39;))
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(926, 20197)
</pre></div></div>
</div>
<p>After metacells are generated, we want to ensure that the distribution of single cells and metacells is qualitatively similar in UMAP space.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_gExpr</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set2&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;UMAP by Cell Types&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-3_28_0.png" src="_images/Tutorial-3_28_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_metacell</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">666</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_metacell</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_metacell</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_metacell</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set2&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;UMAP by Cell Types - Metacells&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-3_29_0.png" src="_images/Tutorial-3_29_0.png" />
</div>
</div>
<p>We notice that there are still two distinct clusters of B cells and of Fibroblast cells, but some B cells are not closer to Stellate cells and Fibroblast cells in UMAP space. We can perform this without a penalty to compare, and then we can also see that the distribution of cells re-used in different metacells has changed due to the penalty.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_metacell_nopenalty</span> <span class="o">=</span> <span class="n">pyviper</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">generateMetacellAnnData</span><span class="p">(</span><span class="n">arg0</span><span class="o">=</span><span class="n">adata_gExpr</span><span class="p">,</span>
                                         <span class="n">method</span><span class="o">=</span><span class="s2">&quot;relative&quot;</span><span class="p">,</span>
                                         <span class="n">cluster_label</span><span class="o">=</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span>
                                         <span class="n">n_metacells</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="c1">#not used</span>
                                         <span class="n">n_metacells_per_cluster</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
                                         <span class="n">n_neighbors</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span>
                                         <span class="n">n_cells_per_metacell</span><span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
                                         <span class="n">experiment_dir_path</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                                         <span class="n">use_decay</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1">#decay factor will not matter now that use_decay is false</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Good quality cells  1000
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "532a9c33e6d64c5da799e4a68d49f5e6", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
Good quality cells  1455
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "90ca73d993b74a1f898b4e1452ef50e4", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
Good quality cells  1277
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0f2f805195754cae9bfc62b60709e28a", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
(255, 20197)
Good quality cells  900
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "88c53e29b66e455b8e6af8114f00d48e", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(200, 20197)
(291, 20197)
(255, 20197)
(180, 20197)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/av2729/opt/anaconda3/envs/single-cell-env-scvi/lib/python3.8/site-packages/scanpy/preprocessing/_normalization.py:197: UserWarning: Some cells have zero counts
  warn(UserWarning(&#39;Some cells have zero counts&#39;))
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(926, 20197)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_metacell_nopenalty</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">666</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_metacell_nopenalty</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_metacell_nopenalty</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_metacell_nopenalty</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;Cell_Type&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;Set2&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;UMAP by Cell Types - Metacells&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-3_32_0.png" src="_images/Tutorial-3_32_0.png" />
</div>
</div>
<p>We observe a similar general distribution of metacells in UMAP space. We can quantitatively identify whether a penalty achieves our desired goal of minimizing overlap across metacells.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#this works if you are working on the source directory of the jupyter notebook</span>
<span class="n">metacell_counts_penalty</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;with_penalty/5/metacell_counts_5.csv&quot;</span><span class="p">)</span>
<span class="n">metacell_counts_nopenalty</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;without_penalty/5/metacell_counts_5.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create and save histograms and bar plots for the cluster</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Histograms of Counts - Cell Types&quot;</span><span class="p">)</span>

<span class="c1"># Plot histograms for &quot;Penalty&quot; and &quot;No Penalty&quot;, 5 and 10 cells per metacell</span>
<span class="k">def</span> <span class="nf">plot_bar</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="p">):</span>
    <span class="n">unique_counts</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">unique_counts</span><span class="o">.</span><span class="n">index</span>
    <span class="n">frequencies</span> <span class="o">=</span> <span class="n">unique_counts</span><span class="o">.</span><span class="n">values</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">frequencies</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Count&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">)</span>

<span class="n">plot_bar</span><span class="p">(</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">metacell_counts_nopenalty</span><span class="p">,</span> <span class="s2">&quot;5 Cells per Metacell - No Penalty&quot;</span><span class="p">)</span>
<span class="n">plot_bar</span><span class="p">(</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">metacell_counts_penalty</span><span class="p">,</span> <span class="s2">&quot;5 Cells per Metacell - Penalty&quot;</span><span class="p">)</span>

<span class="c1"># Add labels</span>
<span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">xlabel</span><span class="o">=</span><span class="s2">&quot;Count&quot;</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>

<span class="c1"># Adjust spacing between subplots</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">rect</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">])</span>
<br/><br/><br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Tutorial-3_35_0.png" src="_images/Tutorial-3_35_0.png" />
</div>
</div>
<p>Using a probability penalty, we have changed from a right tailed distribution to one that peaks at 1, indicating that most of the single cells are selected only once in a metacell.</p>
<p>Now, the metacell datasets can be used to build ARACNe3 networks which can then be used for Py-VIPER analysis.</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Tutorial-2.html" class="btn btn-neutral float-left" title="Tutorial 2 - Inferring Protein Activity from scRNA-seq data from multiple cell populations with the meta-VIPER approach" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, ..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>